{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "022577b2-2952-4605-a9c6-2fedbeffd4b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Track has heterogenous data records. Returning only attributes (columns) shared by all entries.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>585</td>\n",
       "      <td>ENSMMUT00000088937.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>38545</td>\n",
       "      <td>43443</td>\n",
       "      <td>41531</td>\n",
       "      <td>43443</td>\n",
       "      <td>2</td>\n",
       "      <td>38545,43286,</td>\n",
       "      <td>41557,43443,</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMMUG00000064388.1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585</td>\n",
       "      <td>ENSMMUT00000033378.4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>61940</td>\n",
       "      <td>113720</td>\n",
       "      <td>61940</td>\n",
       "      <td>113720</td>\n",
       "      <td>16</td>\n",
       "      <td>61940,62572,63310,69364,69910,88526,88750,9221...</td>\n",
       "      <td>62087,62749,63368,69642,70062,88615,88875,9231...</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMMUG00000023717.4</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>585</td>\n",
       "      <td>ENSMMUT00000069568.2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>62351</td>\n",
       "      <td>113720</td>\n",
       "      <td>62560</td>\n",
       "      <td>113720</td>\n",
       "      <td>15</td>\n",
       "      <td>62351,63310,69364,69910,88526,88750,92219,9239...</td>\n",
       "      <td>62749,63368,69642,70062,88615,88875,92314,9247...</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMMUG00000023717.4</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>585</td>\n",
       "      <td>ENSMMUT00000085572.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>83828</td>\n",
       "      <td>123100</td>\n",
       "      <td>86660</td>\n",
       "      <td>123073</td>\n",
       "      <td>14</td>\n",
       "      <td>83828,87065,88526,88750,92219,92394,92865,1035...</td>\n",
       "      <td>86682,87483,88615,88875,92314,92476,93017,1036...</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMMUG00000023717.4</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,2,0,1,0,1,0,1,0,0,1,0,0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>585</td>\n",
       "      <td>ENSMMUT00000089193.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>83828</td>\n",
       "      <td>122781</td>\n",
       "      <td>86660</td>\n",
       "      <td>122314</td>\n",
       "      <td>14</td>\n",
       "      <td>83828,87065,88526,88750,92219,92394,92865,1035...</td>\n",
       "      <td>86682,87483,88615,88875,92314,92476,93017,1036...</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMMUG00000023717.4</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,2,0,1,0,1,0,1,0,0,1,0,0,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   bin                  name  chrom strand  txStart   txEnd  cdsStart  cdsEnd  \\\n",
       "0  585  ENSMMUT00000088937.1  chr19      -    38545   43443     41531   43443   \n",
       "1  585  ENSMMUT00000033378.4  chr19      -    61940  113720     61940  113720   \n",
       "2  585  ENSMMUT00000069568.2  chr19      -    62351  113720     62560  113720   \n",
       "3  585  ENSMMUT00000085572.1  chr19      -    83828  123100     86660  123073   \n",
       "4  585  ENSMMUT00000089193.1  chr19      -    83828  122781     86660  122314   \n",
       "\n",
       "   exonCount                                         exonStarts  \\\n",
       "0          2                                       38545,43286,   \n",
       "1         16  61940,62572,63310,69364,69910,88526,88750,9221...   \n",
       "2         15  62351,63310,69364,69910,88526,88750,92219,9239...   \n",
       "3         14  83828,87065,88526,88750,92219,92394,92865,1035...   \n",
       "4         14  83828,87065,88526,88750,92219,92394,92865,1035...   \n",
       "\n",
       "                                            exonEnds  score  \\\n",
       "0                                       41557,43443,      0   \n",
       "1  62087,62749,63368,69642,70062,88615,88875,9231...      0   \n",
       "2  62749,63368,69642,70062,88615,88875,92314,9247...      0   \n",
       "3  86682,87483,88615,88875,92314,92476,93017,1036...      0   \n",
       "4  86682,87483,88615,88875,92314,92476,93017,1036...      0   \n",
       "\n",
       "                  name2 cdsStartStat cdsEndStat  \\\n",
       "0  ENSMMUG00000064388.1         cmpl       cmpl   \n",
       "1  ENSMMUG00000023717.4         cmpl       cmpl   \n",
       "2  ENSMMUG00000023717.4         cmpl       cmpl   \n",
       "3  ENSMMUG00000023717.4         cmpl       cmpl   \n",
       "4  ENSMMUG00000023717.4         cmpl       cmpl   \n",
       "\n",
       "                         exonFrames  \n",
       "0                              1,0,  \n",
       "1  0,0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,  \n",
       "2    0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,  \n",
       "3      2,1,2,0,1,0,1,0,1,0,0,1,0,0,  \n",
       "4      2,1,2,0,1,0,1,0,1,0,0,1,0,0,  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import textwrap\n",
    "import re\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "import requests\n",
    "import unicodedata\n",
    "from rapidfuzz import process, fuzz\n",
    "import requests\n",
    "import pandas as pd\n",
    "    \n",
    "def fetch_tracks(assembly):\n",
    "    resp = requests.get(f\"https://api.genome.ucsc.edu/list/tracks?genome={assembly}\")\n",
    "    resp.raise_for_status()\n",
    "    tracks = resp.json()[assembly]\n",
    "    tups = [(name, data['longLabel']) for name, data in tracks.items()]\n",
    "    return sorted(tups, key=lambda t: t[0].upper())\n",
    "\n",
    "def list_ucsc_tracks(assembly=None, label_wrap=80):\n",
    "    search_ucsc_tracks(assembly=assembly,label_wrap=label_wrap)\n",
    "\n",
    "def search_ucsc_tracks(*queries, assembly=None, label_wrap=80):\n",
    "\n",
    "    assert assembly is not None\n",
    "    \n",
    "    def normalize(name):\n",
    "        name = name.lower()\n",
    "        name = unicodedata.normalize(\"NFKD\", name).encode(\"ascii\", \"ignore\").decode()\n",
    "        name = re.sub(r\"[^\\w\\s]\", \"\", name)\n",
    "        return name.strip()\n",
    "\n",
    "    tracks = fetch_tracks(assembly=assembly)\n",
    "    track_names, track_labels = zip(*tracks)\n",
    "\n",
    "    if not queries:  \n",
    "        entries = tracks #names, labels = track_names, track_labels\n",
    "    else:        \n",
    "        normalized_track_names = [normalize(n) for n in track_names]\n",
    "        matches = defaultdict(float)\n",
    "        for query in queries:\n",
    "            for word in query.split():\n",
    "                # match = process.extractOne(query_normalized, normalized_names, scorer=fuzz.WRatio)\n",
    "                # match = process.extractOne(query_normalized, normalized_names, scorer=fuzz.WRatio, score_cutoff=90.0)\n",
    "                search = process.extract(normalize(word), normalized_track_names, scorer=fuzz.WRatio, score_cutoff=80.0, limit=100)\n",
    "                for name, score, index in search:\n",
    "                    matches[(name, index)] += score\n",
    "        \n",
    "        sorted_hits = sorted([(v, k) for k, v in matches.items()], reverse=True)\n",
    "\n",
    "        entries = []\n",
    "        for score, (name, index) in sorted_hits:\n",
    "            entries.append((track_names[index], track_labels[index]))\n",
    "\n",
    "    if entries:\n",
    "        ljust = max([len(e[0]) for e in entries]) + 2\n",
    "        for name, label in entries:\n",
    "            print(name.ljust(ljust) + '\\n'.join(textwrap.wrap(label, width=label_wrap, subsequent_indent=' '*ljust)))\n",
    "\n",
    "def get_ucsc_track(track_name, assembly, chrom=None, start=None, end=None):\n",
    "    url = \"https://api.genome.ucsc.edu/getData/track\"\n",
    "    params = {\"genome\": assembly, \"track\": track_name}\n",
    "    if chrom is not None:\n",
    "        assert chrom.startswith('chr')\n",
    "        params['chrom'] = chrom\n",
    "        if start is not None and end is not None:\n",
    "            params['start'] = str(start)\n",
    "            params['end'] = str(end)            \n",
    "    response = requests.get(url, params=params)\n",
    "    if response.ok:\n",
    "    #    response.raise_for_status()\n",
    "        try:\n",
    "            track_data = response.json().get(track_name, [])\n",
    "        except json.JSONDecodeError:\n",
    "            print(\"Paramters does not represent a valid query\", file=sys.stderr)\n",
    "            return\n",
    "\n",
    "        if type(track_data) is list:\n",
    "            print(\"Track has heterogenous data records. Returning only attributes (columns) shared by all entries.\", file=sys.stderr)\n",
    "            \n",
    "            shared_keys = list(set([k for d in track_data for k in d]))\n",
    "            _track_data = defaultdict(list)\n",
    "            for d in track_data:\n",
    "                for k, v in d.items():\n",
    "                    _track_data[k].append(v)\n",
    "            track_data = _track_data\n",
    "                \n",
    "        try:\n",
    "            return pd.DataFrame(track_data)\n",
    "        except ValueError:\n",
    "            print(df)\n",
    "            raise\n",
    "\n",
    "    \n",
    "df = get_ucsc_track('ensGene', assembly='rheMac10', chrom='chr19')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "06473b10-24be-427a-b189-84da6c999566",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "augustusGene          AUGUSTUS ab initio gene predictions v3.1\n",
      "chainNetHg19          Human (Feb. 2009 (GRCh37/hg19)), Chain and Net Alignments\n",
      "chainNetHg38          Human (Dec. 2013 (GRCh38/hg38)), Chain and Net Alignments\n",
      "chainNetMm10          Mouse (Dec. 2011 (GRCm38/mm10)), Chain and Net Alignments\n",
      "chainNetMm39          Mouse (Jun. 2020 (GRCm39/mm39)), Chain and Net Alignments\n",
      "chainNetNeoSch1       Hawaiian monk seal (Jun. 2017 (ASM220157v1/neoSch1)), Chain and Net Alignments\n",
      "clinvarLift           Human ClinVar variants lifted to Rhesus\n",
      "cpgIslandExt          CpG Islands (Islands < 300 Bases are Light Green)\n",
      "cpgIslandExtUnmasked  CpG Islands on All Sequence (Islands < 300 Bases are Light Green)\n",
      "cytoBandIdeo          Ideogram for Orientation\n",
      "ensGene               Ensembl Genes\n",
      "est                   Rhesus ESTs Including Unspliced\n",
      "evaSnpContainer       Short Genetic Variants from European Variant Archive\n",
      "gap                   Gap Locations\n",
      "gc5BaseBw             GC Percent in 5-Base Windows\n",
      "genscan               Genscan Gene Predictions\n",
      "gold                  Assembly from Fragments\n",
      "HLTOGAannotvHg38v1    TOGA annotations using human/hg38 as reference\n",
      "intronEst             Rhesus ESTs That Have Been Spliced\n",
      "microsat              Microsatellites - Di-nucleotide and Tri-nucleotide Repeats\n",
      "mrna                  Rhesus mRNAs from GenBank\n",
      "nestedRepeats         Fragments of Interrupted Repeats Joined by RepeatMasker ID\n",
      "refSeqComposite       RefSeq genes from NCBI\n",
      "rhesusSNVs            Annotated SNVs from the Rhesus Macaque Sequencing Consortium\n",
      "rmsk                  Repeating Elements by RepeatMasker\n",
      "simpleRepeat          Simple Tandem Repeats by TRF\n",
      "tanDups               Paired identical sequences\n",
      "ucscToINSDC           Accession at INSDC - International Nucleotide Sequence Database Collaboration\n",
      "ucscToRefSeq          RefSeq Accession\n",
      "uniprot               UniProt SwissProt/TrEMBL Protein Annotations\n",
      "windowmaskerSdust     Genomic Intervals Masked by WindowMasker + SDust\n",
      "xenoRefGene           Non-Rhesus RefSeq Genes\n"
     ]
    }
   ],
   "source": [
    "list_ucsc_tracks(assembly='rheMac10')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "58bac6c5-68a7-44a3-ac81-13fa57b86119",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       bin                  name  chrom strand   txStart     txEnd  cdsStart  \\\n",
      "0      585  ENSMMUT00000088937.1  chr19      -     38545     43443     41531   \n",
      "1      585  ENSMMUT00000033378.4  chr19      -     61940    113720     61940   \n",
      "2      585  ENSMMUT00000069568.2  chr19      -     62351    113720     62560   \n",
      "3      585  ENSMMUT00000085572.1  chr19      -     83828    123100     86660   \n",
      "4      585  ENSMMUT00000089193.1  chr19      -     83828    122781     86660   \n",
      "...    ...                   ...    ...    ...       ...       ...       ...   \n",
      "3510  1029  ENSMMUT00000099271.1  chr19      +  58249911  58271005  58249911   \n",
      "3511  1029  ENSMMUT00000089503.1  chr19      +  58249911  58271005  58249911   \n",
      "3512  1029  ENSMMUT00000038349.3  chr19      -  58285362  58312850  58285875   \n",
      "3513  1029  ENSMMUT00000085828.1  chr19      -  58285362  58297696  58285875   \n",
      "3514  1029  ENSMMUT00000030783.4  chr19      -  58285369  58312847  58285875   \n",
      "\n",
      "        cdsEnd  exonCount                                         exonStarts  \\\n",
      "0        43443          2                                       38545,43286,   \n",
      "1       113720         16  61940,62572,63310,69364,69910,88526,88750,9221...   \n",
      "2       113720         15  62351,63310,69364,69910,88526,88750,92219,9239...   \n",
      "3       123073         14  83828,87065,88526,88750,92219,92394,92865,1035...   \n",
      "4       122314         14  83828,87065,88526,88750,92219,92394,92865,1035...   \n",
      "...        ...        ...                                                ...   \n",
      "3510  58271005         24  58249911,58251230,58253737,58254385,58255113,5...   \n",
      "3511  58271005         19  58249911,58251230,58253737,58254977,58258516,5...   \n",
      "3512  58312841          9  58285362,58286151,58286665,58287678,58290073,5...   \n",
      "3513  58297599          9  58285362,58286151,58286665,58287678,58290073,5...   \n",
      "3514  58312841          9  58285369,58286151,58286665,58287678,58306860,5...   \n",
      "\n",
      "                                               exonEnds  score  \\\n",
      "0                                          41557,43443,      0   \n",
      "1     62087,62749,63368,69642,70062,88615,88875,9231...      0   \n",
      "2     62749,63368,69642,70062,88615,88875,92314,9247...      0   \n",
      "3     86682,87483,88615,88875,92314,92476,93017,1036...      0   \n",
      "4     86682,87483,88615,88875,92314,92476,93017,1036...      0   \n",
      "...                                                 ...    ...   \n",
      "3510  58249924,58251247,58253782,58254425,58255164,5...      0   \n",
      "3511  58249924,58251247,58253784,58255000,58258530,5...      0   \n",
      "3512  58286052,58286204,58286767,58287729,58290367,5...      0   \n",
      "3513  58286052,58286204,58286767,58287729,58290367,5...      0   \n",
      "3514  58286052,58286204,58286767,58287729,58307154,5...      0   \n",
      "\n",
      "                     name2 cdsStartStat cdsEndStat  \\\n",
      "0     ENSMMUG00000064388.1         cmpl       cmpl   \n",
      "1     ENSMMUG00000023717.4         cmpl       cmpl   \n",
      "2     ENSMMUG00000023717.4         cmpl       cmpl   \n",
      "3     ENSMMUG00000023717.4         cmpl       cmpl   \n",
      "4     ENSMMUG00000023717.4         cmpl       cmpl   \n",
      "...                    ...          ...        ...   \n",
      "3510  ENSMMUG00000054609.1       incmpl     incmpl   \n",
      "3511  ENSMMUG00000054609.1       incmpl     incmpl   \n",
      "3512  ENSMMUG00000009456.4         cmpl       cmpl   \n",
      "3513  ENSMMUG00000009456.4         cmpl       cmpl   \n",
      "3514  ENSMMUG00000009456.4         cmpl       cmpl   \n",
      "\n",
      "                                            exonFrames  \n",
      "0                                                 1,0,  \n",
      "1                     0,0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,  \n",
      "2                       0,2,0,1,2,0,1,0,1,0,1,0,0,1,0,  \n",
      "3                         2,1,2,0,1,0,1,0,1,0,0,1,0,0,  \n",
      "4                         2,1,2,0,1,0,1,0,1,0,0,1,0,0,  \n",
      "...                                                ...  \n",
      "3510  0,1,0,0,1,1,0,0,2,1,1,1,2,2,0,1,0,0,0,1,1,1,2,1,  \n",
      "3511            0,1,0,2,1,0,1,2,0,2,2,1,1,1,2,1,1,2,1,  \n",
      "3512                                0,1,1,1,1,1,1,1,0,  \n",
      "3513                                0,1,1,1,1,1,1,1,0,  \n",
      "3514                                0,1,1,1,1,1,1,1,0,  \n",
      "\n",
      "[3515 rows x 16 columns]\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "All arrays must be of the same length",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[209]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mget_ucsc_track\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mensGene\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43massembly\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mrheMac10\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[207]\u001b[39m\u001b[32m, line 76\u001b[39m, in \u001b[36mget_ucsc_track\u001b[39m\u001b[34m(track_name, assembly, chrom, start, end)\u001b[39m\n\u001b[32m     74\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[32m     75\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m76\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mDataFrame\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrack_data\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     77\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[32m     78\u001b[39m     \u001b[38;5;28mprint\u001b[39m(df)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/pandas/core/frame.py:778\u001b[39m, in \u001b[36mDataFrame.__init__\u001b[39m\u001b[34m(self, data, index, columns, dtype, copy)\u001b[39m\n\u001b[32m    772\u001b[39m     mgr = \u001b[38;5;28mself\u001b[39m._init_mgr(\n\u001b[32m    773\u001b[39m         data, axes={\u001b[33m\"\u001b[39m\u001b[33mindex\u001b[39m\u001b[33m\"\u001b[39m: index, \u001b[33m\"\u001b[39m\u001b[33mcolumns\u001b[39m\u001b[33m\"\u001b[39m: columns}, dtype=dtype, copy=copy\n\u001b[32m    774\u001b[39m     )\n\u001b[32m    776\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, \u001b[38;5;28mdict\u001b[39m):\n\u001b[32m    777\u001b[39m     \u001b[38;5;66;03m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m778\u001b[39m     mgr = \u001b[43mdict_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmanager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    779\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, ma.MaskedArray):\n\u001b[32m    780\u001b[39m     \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnumpy\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mma\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m mrecords\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/pandas/core/internals/construction.py:503\u001b[39m, in \u001b[36mdict_to_mgr\u001b[39m\u001b[34m(data, index, columns, dtype, typ, copy)\u001b[39m\n\u001b[32m    499\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    500\u001b[39m         \u001b[38;5;66;03m# dtype check to exclude e.g. range objects, scalars\u001b[39;00m\n\u001b[32m    501\u001b[39m         arrays = [x.copy() \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(x, \u001b[33m\"\u001b[39m\u001b[33mdtype\u001b[39m\u001b[33m\"\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m arrays]\n\u001b[32m--> \u001b[39m\u001b[32m503\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43marrays_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/pandas/core/internals/construction.py:114\u001b[39m, in \u001b[36marrays_to_mgr\u001b[39m\u001b[34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[39m\n\u001b[32m    111\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m verify_integrity:\n\u001b[32m    112\u001b[39m     \u001b[38;5;66;03m# figure out the index, if necessary\u001b[39;00m\n\u001b[32m    113\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m index \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m114\u001b[39m         index = \u001b[43m_extract_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    115\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    116\u001b[39m         index = ensure_index(index)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/pandas/core/internals/construction.py:677\u001b[39m, in \u001b[36m_extract_index\u001b[39m\u001b[34m(data)\u001b[39m\n\u001b[32m    675\u001b[39m lengths = \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mset\u001b[39m(raw_lengths))\n\u001b[32m    676\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(lengths) > \u001b[32m1\u001b[39m:\n\u001b[32m--> \u001b[39m\u001b[32m677\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mAll arrays must be of the same length\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    679\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m have_dicts:\n\u001b[32m    680\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m    681\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mMixing dicts with non-Series may lead to ambiguous ordering.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    682\u001b[39m     )\n",
      "\u001b[31mValueError\u001b[39m: All arrays must be of the same length"
     ]
    }
   ],
   "source": [
    "get_ucsc_track('ensGene', assembly='rheMac10')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "f6509c8d-8a57-44b4-b2c4-d580a28b25bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>chrUn_NW_021162293v1</th>\n",
       "      <th>chrUn_NW_021162233v1</th>\n",
       "      <th>chrUn_NW_021162201v1</th>\n",
       "      <th>chrUn_NW_021162242v1</th>\n",
       "      <th>chrUn_NW_021162064v1</th>\n",
       "      <th>chrUn_NW_021162018v1</th>\n",
       "      <th>chrUn_NW_021162284v1</th>\n",
       "      <th>chrUn_NW_021162073v1</th>\n",
       "      <th>chrUn_NW_021162254v1</th>\n",
       "      <th>...</th>\n",
       "      <th>chr11</th>\n",
       "      <th>chr9</th>\n",
       "      <th>chr8</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chr7</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'chrom': 'chr1', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162293v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162233v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162201v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162242v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162064v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162018v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162284v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162073v1', 'chromStart'...</td>\n",
       "      <td>{'chrom': 'chrUn_NW_021162254v1', 'chromStart'...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'chrom': 'chr11', 'chromStart': 0, 'chromEnd'...</td>\n",
       "      <td>{'chrom': 'chr9', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr8', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chrX', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr7', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr4', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr6', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr3', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr5', 'chromStart': 0, 'chromEnd':...</td>\n",
       "      <td>{'chrom': 'chr2', 'chromStart': 0, 'chromEnd':...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 2939 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                chr1  \\\n",
       "0  {'chrom': 'chr1', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                chrUn_NW_021162293v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162293v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162233v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162233v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162201v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162201v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162242v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162242v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162064v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162064v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162018v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162018v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162284v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162284v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162073v1  \\\n",
       "0  {'chrom': 'chrUn_NW_021162073v1', 'chromStart'...   \n",
       "\n",
       "                                chrUn_NW_021162254v1  ...  \\\n",
       "0  {'chrom': 'chrUn_NW_021162254v1', 'chromStart'...  ...   \n",
       "\n",
       "                                               chr11  \\\n",
       "0  {'chrom': 'chr11', 'chromStart': 0, 'chromEnd'...   \n",
       "\n",
       "                                                chr9  \\\n",
       "0  {'chrom': 'chr9', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr8  \\\n",
       "0  {'chrom': 'chr8', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chrX  \\\n",
       "0  {'chrom': 'chrX', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr7  \\\n",
       "0  {'chrom': 'chr7', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr4  \\\n",
       "0  {'chrom': 'chr4', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr6  \\\n",
       "0  {'chrom': 'chr6', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr3  \\\n",
       "0  {'chrom': 'chr3', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr5  \\\n",
       "0  {'chrom': 'chr5', 'chromStart': 0, 'chromEnd':...   \n",
       "\n",
       "                                                chr2  \n",
       "0  {'chrom': 'chr2', 'chromStart': 0, 'chromEnd':...  \n",
       "\n",
       "[1 rows x 2939 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_ucsc_track('ucscToRefSeq', assembly='rheMac10')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "fb84c49f-8a76-4d74-92fd-9db0a7fb75c0",
   "metadata": {},
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "404 Client Error: Not Found for url: https://data.4dnucleome.org/search/?type=File&file_format=bedpe&output_type=loop+list&status=released&assay_title=Hi-C&biosample_ontology.term_name=GM12878&limit=5&format=json",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mHTTPError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[133]\u001b[39m\u001b[32m, line 22\u001b[39m\n\u001b[32m     19\u001b[39m headers = {\u001b[33m\"\u001b[39m\u001b[33maccept\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mapplication/json\u001b[39m\u001b[33m\"\u001b[39m}\n\u001b[32m     21\u001b[39m r = requests.get(url, headers=headers, params=params, auth=HTTPBasicAuth(key, secret))\n\u001b[32m---> \u001b[39m\u001b[32m22\u001b[39m \u001b[43mr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mraise_for_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     23\u001b[39m results = r.json()[\u001b[33m\"\u001b[39m\u001b[33m@graph\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m     25\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m results:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/requests/models.py:1026\u001b[39m, in \u001b[36mResponse.raise_for_status\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1021\u001b[39m     http_error_msg = (\n\u001b[32m   1022\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.status_code\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m Server Error: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mreason\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m for url: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.url\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m   1023\u001b[39m     )\n\u001b[32m   1025\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m http_error_msg:\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m HTTPError(http_error_msg, response=\u001b[38;5;28mself\u001b[39m)\n",
      "\u001b[31mHTTPError\u001b[39m: 404 Client Error: Not Found for url: https://data.4dnucleome.org/search/?type=File&file_format=bedpe&output_type=loop+list&status=released&assay_title=Hi-C&biosample_ontology.term_name=GM12878&limit=5&format=json"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from requests.auth import HTTPBasicAuth\n",
    "\n",
    "\n",
    "key = '4TB2XFCS'\n",
    "secret = 'prgc6t46emkfuknd'\n",
    "\n",
    "url = \"https://data.4dnucleome.org/search/\"\n",
    "params = {\n",
    "    \"type\": \"File\",\n",
    "    \"file_format\": \"bedpe\",\n",
    "    \"output_type\": \"loop list\",\n",
    "    \"status\": \"released\",\n",
    "    \"assay_title\": \"Hi-C\",\n",
    "    \"biosample_ontology.term_name\": \"GM12878\",\n",
    "    \"limit\": \"5\",\n",
    "    \"format\": \"json\"\n",
    "}\n",
    "headers = {\"accept\": \"application/json\"}\n",
    "\n",
    "r = requests.get(url, headers=headers, params=params, auth=HTTPBasicAuth(key, secret))\n",
    "r.raise_for_status()\n",
    "results = r.json()[\"@graph\"]\n",
    "\n",
    "for file in results:\n",
    "    print(file[\"accession\"], file[\"href\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "8266a3ae-9375-4956-b835-3b0017e79a7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1dec559c8bfe48a999569818245e0cba",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/211524 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import requests\n",
    "from requests.auth import HTTPBasicAuth\n",
    "import json\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "def download_4dn(identifier, dowload_dir=os.getcwd(), pgbar=False):\n",
    "    \n",
    "    def download_file(url, dowload_dir=dowload_dir):\n",
    "        if not os.path.exists(dowload_dir):\n",
    "            os.makedirs(dowload_dir)\n",
    "        file_name = url.split('/')[-1]\n",
    "        with requests.get(url, stream=True) as r:\n",
    "            r.raise_for_status()\n",
    "            with open(file_name, 'wb') as f:\n",
    "                if pgbar:\n",
    "                    pbar = tqdm(total=int(r.headers['Content-Length']))\n",
    "                for chunk in r.iter_content(chunk_size=8192):\n",
    "                    if chunk:\n",
    "                        f.write(chunk)         \n",
    "                        if pgbar:\n",
    "                            pbar.update(len(chunk))\n",
    "\n",
    "    url = f\"https://data.4dnucleome.org/ga4gh/drs/v1/objects/{identifier}/access/https\"\n",
    "    response = requests.get(url, auth=HTTPBasicAuth(key, secret))\n",
    "    if not response.ok:\n",
    "        assert 0\n",
    "    info = response.json()    \n",
    "    download_file(info[\"url\"])\n",
    "\n",
    "download_4dn('4DNFIA85JYD7', pgbar=True)\n",
    "\n",
    "\n",
    "# 4DNESUB35TII \n",
    "# 4DNFI8HHP7VN\n",
    "# 4DNFIDWGZLHX\n",
    "# 4DNFIWCH6ADP\n",
    "# 4DNFII3DT59U\n",
    "# 4DNFILKF6IA2\n",
    "# 4DNFIA85JYD7\n",
    "# 4DNEXOTMVPJN \n",
    "# 4DNFIMPBTBKI\n",
    "# 4DNFIEBHQU8O"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a36859da-e4ff-47fd-a382-e037eaa34ffe",
   "metadata": {},
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "404 Client Error: Not Found for url: https://data.4dnucleome.org/search/?type=File&file_format=bedpe&output_type=loop+list&status=released&assay_term_name=Hi-C&biosample_ontology.term_name=GM12878&limit=10&format=json",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mHTTPError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[24]\u001b[39m\u001b[32m, line 20\u001b[39m\n\u001b[32m     17\u001b[39m     \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFound \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(results)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m loop files for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcell_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m     18\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m results\n\u001b[32m---> \u001b[39m\u001b[32m20\u001b[39m \u001b[43mfetch_loop_files\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[24]\u001b[39m\u001b[32m, line 15\u001b[39m, in \u001b[36mfetch_loop_files\u001b[39m\u001b[34m(cell_type, limit)\u001b[39m\n\u001b[32m      4\u001b[39m params = {\n\u001b[32m      5\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mtype\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mFile\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      6\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mfile_format\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mbedpe\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     12\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mformat\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mjson\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m     13\u001b[39m }\n\u001b[32m     14\u001b[39m r = requests.get(url, headers=HEADERS, params=params, auth=HTTPBasicAuth(key, secret))\n\u001b[32m---> \u001b[39m\u001b[32m15\u001b[39m \u001b[43mr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mraise_for_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     16\u001b[39m results = r.json()[\u001b[33m\"\u001b[39m\u001b[33m@graph\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m     17\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFound \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(results)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m loop files for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcell_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/geneinfo/.pixi/envs/default/lib/python3.13/site-packages/requests/models.py:1026\u001b[39m, in \u001b[36mResponse.raise_for_status\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1021\u001b[39m     http_error_msg = (\n\u001b[32m   1022\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.status_code\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m Server Error: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mreason\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m for url: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.url\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m   1023\u001b[39m     )\n\u001b[32m   1025\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m http_error_msg:\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m HTTPError(http_error_msg, response=\u001b[38;5;28mself\u001b[39m)\n",
      "\u001b[31mHTTPError\u001b[39m: 404 Client Error: Not Found for url: https://data.4dnucleome.org/search/?type=File&file_format=bedpe&output_type=loop+list&status=released&assay_term_name=Hi-C&biosample_ontology.term_name=GM12878&limit=10&format=json"
     ]
    }
   ],
   "source": [
    "def fetch_loop_files(cell_type=\"GM12878\", limit=10):\n",
    "    \"\"\"Query 4DN for Hi-C loop files in the given human cell type\"\"\"\n",
    "    url = f\"{API_BASE}/search/\"\n",
    "    params = {\n",
    "        \"type\": \"File\",\n",
    "        \"file_format\": \"bedpe\",\n",
    "        \"output_type\": \"loop list\",\n",
    "        \"status\": \"released\",\n",
    "        \"assay_term_name\": \"Hi-C\",\n",
    "        \"biosample_ontology.term_name\": cell_type,\n",
    "        \"limit\": str(limit),\n",
    "        \"format\": \"json\"\n",
    "    }\n",
    "    r = requests.get(url, headers=HEADERS, params=params, auth=HTTPBasicAuth(key, secret))\n",
    "    r.raise_for_status()\n",
    "    results = r.json()[\"@graph\"]\n",
    "    print(f\"Found {len(results)} loop files for {cell_type}\")\n",
    "    return results\n",
    "\n",
    "fetch_loop_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8101610-3c28-4dd8-855b-986570d052fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {
     "0004eac710d54cd3bde1206153902a96": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "HTMLStyleModel",
      "state": {
       "description_width": "",
       "font_size": null,
       "text_color": null
      }
     },
     "1dec559c8bfe48a999569818245e0cba": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "HBoxModel",
      "state": {
       "children": [
        "IPY_MODEL_dc1dc12baa2f43498840da1dcfe3fcff",
        "IPY_MODEL_aea53b76d9914591b9a943120f5f9328",
        "IPY_MODEL_d571ab0b4ce749acb019cd1221df8703"
       ],
       "layout": "IPY_MODEL_9a1b59bd4e4d461ea8c2ebe0c5b44723"
      }
     },
     "25c0fbd1d546457288feec8c90115a41": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "HTMLStyleModel",
      "state": {
       "description_width": "",
       "font_size": null,
       "text_color": null
      }
     },
     "6c7bbd1bc5a045d3ae332d187228175d": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "2.0.0",
      "model_name": "LayoutModel",
      "state": {}
     },
     "7f3ae4be29ec440587508c6670e73c0a": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "2.0.0",
      "model_name": "LayoutModel",
      "state": {}
     },
     "9a1b59bd4e4d461ea8c2ebe0c5b44723": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "2.0.0",
      "model_name": "LayoutModel",
      "state": {}
     },
     "aea53b76d9914591b9a943120f5f9328": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "FloatProgressModel",
      "state": {
       "bar_style": "success",
       "layout": "IPY_MODEL_e7bc93dac8264d6c903eae8ffcf7d5bf",
       "max": 211524,
       "style": "IPY_MODEL_dbe99754c5894fb48ae08c6d774adfd0",
       "value": 211524
      }
     },
     "d571ab0b4ce749acb019cd1221df8703": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "HTMLModel",
      "state": {
       "layout": "IPY_MODEL_7f3ae4be29ec440587508c6670e73c0a",
       "style": "IPY_MODEL_0004eac710d54cd3bde1206153902a96",
       "value": " 211524/211524 [1:59:53&lt;00:00, 482031.87it/s]"
      }
     },
     "dbe99754c5894fb48ae08c6d774adfd0": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "ProgressStyleModel",
      "state": {
       "description_width": ""
      }
     },
     "dc1dc12baa2f43498840da1dcfe3fcff": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "2.0.0",
      "model_name": "HTMLModel",
      "state": {
       "layout": "IPY_MODEL_6c7bbd1bc5a045d3ae332d187228175d",
       "style": "IPY_MODEL_25c0fbd1d546457288feec8c90115a41",
       "value": "100%"
      }
     },
     "e7bc93dac8264d6c903eae8ffcf7d5bf": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "2.0.0",
      "model_name": "LayoutModel",
      "state": {}
     }
    },
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
