<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Ideograms â€“ geneinfo</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/string_nets.html" rel="next">
<link href="../pages/gene_plots.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-5c67a7444eebc356e54fdce0a63e8750.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../api/_styles-quartodoc.css">
<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">geneinfo</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../api/"> 
<span class="menu-text">API reference</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../gallery/"> 
<span class="menu-text">Gallery</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/munch-group/geneinfo/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/gene_plots.html">Visualization</a></li><li class="breadcrumb-item"><a href="../pages/chrom_ideograms.html">Ideograms</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Information</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene_ids.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ID conversion</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gene information</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Visualization</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene_plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gene context</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/chrom_ideograms.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Ideograms</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/string_nets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">String networks</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/go_graphs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GO graphs</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Enrichment</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/go_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GO information</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/go_enrichment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GO enrichment</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Utilities</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene_lists.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gene lists</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/cashing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Caching</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#dummy-sample-data" id="toc-dummy-sample-data" class="nav-link active" data-scroll-target="#dummy-sample-data">Dummy sample data</a></li>
  <li><a href="#all-chromosome-ideogram" id="toc-all-chromosome-ideogram" class="nav-link" data-scroll-target="#all-chromosome-ideogram">All chromosome ideogram</a></li>
  <li><a href="#single-chromosome-ideogram" id="toc-single-chromosome-ideogram" class="nav-link" data-scroll-target="#single-chromosome-ideogram">Single chromosome ideogram</a></li>
  <li><a href="#the-coordinate-system" id="toc-the-coordinate-system" class="nav-link" data-scroll-target="#the-coordinate-system">The coordinate system</a></li>
  <li><a href="#ideogram-size-and-plament" id="toc-ideogram-size-and-plament" class="nav-link" data-scroll-target="#ideogram-size-and-plament">Ideogram size and plament</a></li>
  <li><a href="#genomic-segment-annotation" id="toc-genomic-segment-annotation" class="nav-link" data-scroll-target="#genomic-segment-annotation">Genomic segment annotation</a></li>
  <li><a href="#conditional-ideogram-color" id="toc-conditional-ideogram-color" class="nav-link" data-scroll-target="#conditional-ideogram-color">Conditional ideogram color</a></li>
  <li><a href="#context-plots" id="toc-context-plots" class="nav-link" data-scroll-target="#context-plots">Context plots</a></li>
  <li><a href="#labelled-positions" id="toc-labelled-positions" class="nav-link" data-scroll-target="#labelled-positions">Labelled positions</a></li>
  <li><a href="#combinatorial-highlighting" id="toc-combinatorial-highlighting" class="nav-link" data-scroll-target="#combinatorial-highlighting">Combinatorial highlighting</a></li>
  <li><a href="#additional-axes" id="toc-additional-axes" class="nav-link" data-scroll-target="#additional-axes">Additional axes</a></li>
  <li><a href="#floating-annotations" id="toc-floating-annotations" class="nav-link" data-scroll-target="#floating-annotations">Floating annotations</a></li>
  <li><a href="#zooms" id="toc-zooms" class="nav-link" data-scroll-target="#zooms">Zooms</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/gene_plots.html">Visualization</a></li><li class="breadcrumb-item"><a href="../pages/chrom_ideograms.html">Ideograms</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Ideograms</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="cell-2" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:15.763367Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:15.763193Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:16.564224Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:16.563866Z&quot;}" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random, re</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> random <span class="im">import</span> randint</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>random.seed(<span class="dv">7</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy  <span class="im">as</span> np</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.axes <span class="im">import</span> Axes</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>sns.set_style(<span class="st">'white'</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># %config InlineBackend.figure_format = 'retina'</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>config InlineBackend.figure_format <span class="op">=</span> <span class="st">'svg'</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo.utils <span class="im">as</span> utils</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo.information <span class="im">as</span> gi</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo.plot <span class="im">as</span> gplt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-3" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:16.566217Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:16.566012Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:16.663140Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:16.662829Z&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>gene_lists <span class="op">=</span> utils.GeneListCollection(<span class="st">'google_sheet.csv'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>gene_lists</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">label</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>all_npx</strong></td>
<td style="text-align: left;">all NPX genes</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>neuron_genome_proteome</strong></td>
<td style="text-align: left;">Neuron proteome (whole genome)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>neuron_npx_proteome</strong></td>
<td style="text-align: left;">NPX neuron proteome</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>matos_neuron</strong></td>
<td style="text-align: left;">Matos neuron genes</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>matos_common</strong></td>
<td style="text-align: left;">Matos neuron genes expressed in spermatids</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>hpa_brain_prot</strong></td>
<td style="text-align: left;">Brain genes form Human protein atlas</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>meritxell_spermatid_expr</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>intersection(nDEG, matos_neuron)</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>union(H,matos_common)</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>mult_copy</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>all_post_mei_expr</strong></td>
<td style="text-align: left;">Multi copy genes</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>msci</strong></td>
<td style="text-align: left;">all post.mei.expr. X genes</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>msci_esc</strong></td>
<td style="text-align: left;">MSCI genes</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>my_primate_codeml</strong></td>
<td style="text-align: left;">MSCI-escaping</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>cDEG</strong></td>
<td style="text-align: left;">Primate codeml site model</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>nDEG</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>primate_ampl_multi</strong></td>
<td style="text-align: left;">cDEG</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>gametologs</strong></td>
<td style="text-align: left;">nDEG</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>xi</strong></td>
<td style="text-align: left;">Ampliconic or multicopy in some primates</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>xi_escape</strong></td>
<td style="text-align: left;">Gametologs in some primates</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>xi_uncertain</strong></td>
<td style="text-align: left;">Subject to somatic X inactivation (XI)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>xi_any_evidence</strong></td>
<td style="text-align: left;">Escaping somatic X inactivation (XI)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>expr_mod_xi_copynr_fibrobl</strong></td>
<td style="text-align: left;">X genes uncertain XI status</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>expr_mod_xi_copynr_lcl</strong></td>
<td style="text-align: left;">Union of certain and uncertain XI status</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>reg_sa_pheno</strong></td>
<td style="text-align: left;">Expression modulated by Xi copy number (Fibroblast)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>ari_relate_EUR</strong></td>
<td style="text-align: left;">Expression modulated by Xi copy number (LCL)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>ari_relate_ASIA</strong></td>
<td style="text-align: left;">X-chromosome genes associated with regional SA phenotypes</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>ari_relate_AFR</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>ari_nonPUR</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>ari_relate_PUR</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>ari_all</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>candidates</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>ech75_regions</strong></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>ech90_regions</strong></td>
<td style="text-align: left;">Candidate for explaining human and baboon sweeps</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>accel_reg_simiiformes_br</strong></td>
<td style="text-align: left;">Overlapping human sweeps (75% ECH regions)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>sfari_all_conf</strong></td>
<td style="text-align: left;">Overlapping human sweeps (90% ECH regions)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>intel_seiz_lang</strong></td>
<td style="text-align: left;">Genes with accelerated regions Simiiformes branch</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>intelect_disabil</strong></td>
<td style="text-align: left;">X SFARI genes (including all confidence levels)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>matos_common_no_xy_gametologs</strong></td>
<td style="text-align: left;">Intellectual disability or Seizures or Language impairment</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code></code></pre>
</div>
</div>
<p><strong>geneinfo</strong> lets you visualize data in a chromosome or genome context.</p>
<section id="dummy-sample-data" class="level2">
<h2 class="anchored" data-anchor-id="dummy-sample-data">Dummy sample data</h2>
<div id="cell-5" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:16.685625Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:16.685390Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:16.734117Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:16.733820Z&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> utils.dummy_data()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">y</th>
<th data-quarto-table-cell-role="th">chrom</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">pos</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>5.019283e+05</td>
<td>0.063933</td>
<td>chr1</td>
<td>0.000000e+00</td>
<td>5.019283e+05</td>
<td>0.000000e+00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>5.019283e+05</td>
<td>0.063933</td>
<td>chr1</td>
<td>0.000000e+00</td>
<td>5.019283e+05</td>
<td>5.019283e+05</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>1.003857e+06</td>
<td>0.128872</td>
<td>chr1</td>
<td>5.019283e+05</td>
<td>1.003857e+06</td>
<td>5.019283e+05</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>1.003857e+06</td>
<td>0.128872</td>
<td>chr1</td>
<td>5.019283e+05</td>
<td>1.003857e+06</td>
<td>1.003857e+06</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>1.505785e+06</td>
<td>0.194514</td>
<td>chr1</td>
<td>1.003857e+06</td>
<td>1.505785e+06</td>
<td>1.003857e+06</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-6" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:16.735685Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:16.735564Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:16.749494Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:16.749215Z&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>segments <span class="op">=</span> utils.dummy_segments()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>segments[:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>[('chr1', 10065165, 12961789),
 ('chr1', 15568967, 18751672),
 ('chr1', 19444467, 23071285),
 ('chr1', 24350589, 25267841),
 ('chr1', 40493267, 57632604),
 ('chr1', 64602482, 86928195),
 ('chr1', 98163871, 105984624),
 ('chr1', 112252233, 113956002),
 ('chr1', 116405877, 136213743),
 ('chr1', 143849730, 147920620)]</code></pre>
</div>
</div>
</section>
<section id="all-chromosome-ideogram" class="level2">
<h2 class="anchored" data-anchor-id="all-chromosome-ideogram">All chromosome ideogram</h2>
<div id="cell-8" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:16.751253Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:16.751148Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:16.763374Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:16.763100Z&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo.plot <span class="im">as</span> gplt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Segments:</p>
<div id="cell-10" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:16.765076Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:16.764980Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:19.082885Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:19.082081Z&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-8-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="chrom_ideograms_files/figure-html/cell-8-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-11" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:19.085345Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:19.085199Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:19.876757Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:19.876450Z&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>gi.list_assemblies()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style type="text/css">
#T_6533d th {
  text-align: left;
}
#T_6533d_row0_col0, #T_6533d_row0_col1, #T_6533d_row0_col2, #T_6533d_row1_col0, #T_6533d_row1_col1, #T_6533d_row1_col2, #T_6533d_row2_col0, #T_6533d_row2_col1, #T_6533d_row2_col2, #T_6533d_row3_col0, #T_6533d_row3_col1, #T_6533d_row3_col2, #T_6533d_row4_col0, #T_6533d_row4_col1, #T_6533d_row4_col2, #T_6533d_row5_col0, #T_6533d_row5_col1, #T_6533d_row5_col2, #T_6533d_row6_col0, #T_6533d_row6_col1, #T_6533d_row6_col2, #T_6533d_row7_col0, #T_6533d_row7_col1, #T_6533d_row7_col2, #T_6533d_row8_col0, #T_6533d_row8_col1, #T_6533d_row8_col2, #T_6533d_row9_col0, #T_6533d_row9_col1, #T_6533d_row9_col2, #T_6533d_row10_col0, #T_6533d_row10_col1, #T_6533d_row10_col2, #T_6533d_row11_col0, #T_6533d_row11_col1, #T_6533d_row11_col2, #T_6533d_row12_col0, #T_6533d_row12_col1, #T_6533d_row12_col2, #T_6533d_row13_col0, #T_6533d_row13_col1, #T_6533d_row13_col2, #T_6533d_row14_col0, #T_6533d_row14_col1, #T_6533d_row14_col2, #T_6533d_row15_col0, #T_6533d_row15_col1, #T_6533d_row15_col2, #T_6533d_row16_col0, #T_6533d_row16_col1, #T_6533d_row16_col2, #T_6533d_row17_col0, #T_6533d_row17_col1, #T_6533d_row17_col2, #T_6533d_row18_col0, #T_6533d_row18_col1, #T_6533d_row18_col2, #T_6533d_row19_col0, #T_6533d_row19_col1, #T_6533d_row19_col2, #T_6533d_row20_col0, #T_6533d_row20_col1, #T_6533d_row20_col2, #T_6533d_row21_col0, #T_6533d_row21_col1, #T_6533d_row21_col2, #T_6533d_row22_col0, #T_6533d_row22_col1, #T_6533d_row22_col2, #T_6533d_row23_col0, #T_6533d_row23_col1, #T_6533d_row23_col2, #T_6533d_row24_col0, #T_6533d_row24_col1, #T_6533d_row24_col2, #T_6533d_row25_col0, #T_6533d_row25_col1, #T_6533d_row25_col2, #T_6533d_row26_col0, #T_6533d_row26_col1, #T_6533d_row26_col2, #T_6533d_row27_col0, #T_6533d_row27_col1, #T_6533d_row27_col2, #T_6533d_row28_col0, #T_6533d_row28_col1, #T_6533d_row28_col2, #T_6533d_row29_col0, #T_6533d_row29_col1, #T_6533d_row29_col2, #T_6533d_row30_col0, #T_6533d_row30_col1, #T_6533d_row30_col2, #T_6533d_row31_col0, #T_6533d_row31_col1, #T_6533d_row31_col2, #T_6533d_row32_col0, #T_6533d_row32_col1, #T_6533d_row32_col2, #T_6533d_row33_col0, #T_6533d_row33_col1, #T_6533d_row33_col2, #T_6533d_row34_col0, #T_6533d_row34_col1, #T_6533d_row34_col2, #T_6533d_row35_col0, #T_6533d_row35_col1, #T_6533d_row35_col2, #T_6533d_row36_col0, #T_6533d_row36_col1, #T_6533d_row36_col2, #T_6533d_row37_col0, #T_6533d_row37_col1, #T_6533d_row37_col2, #T_6533d_row38_col0, #T_6533d_row38_col1, #T_6533d_row38_col2, #T_6533d_row39_col0, #T_6533d_row39_col1, #T_6533d_row39_col2, #T_6533d_row40_col0, #T_6533d_row40_col1, #T_6533d_row40_col2, #T_6533d_row41_col0, #T_6533d_row41_col1, #T_6533d_row41_col2, #T_6533d_row42_col0, #T_6533d_row42_col1, #T_6533d_row42_col2, #T_6533d_row43_col0, #T_6533d_row43_col1, #T_6533d_row43_col2, #T_6533d_row44_col0, #T_6533d_row44_col1, #T_6533d_row44_col2, #T_6533d_row45_col0, #T_6533d_row45_col1, #T_6533d_row45_col2, #T_6533d_row46_col0, #T_6533d_row46_col1, #T_6533d_row46_col2, #T_6533d_row47_col0, #T_6533d_row47_col1, #T_6533d_row47_col2, #T_6533d_row48_col0, #T_6533d_row48_col1, #T_6533d_row48_col2, #T_6533d_row49_col0, #T_6533d_row49_col1, #T_6533d_row49_col2, #T_6533d_row50_col0, #T_6533d_row50_col1, #T_6533d_row50_col2, #T_6533d_row51_col0, #T_6533d_row51_col1, #T_6533d_row51_col2, #T_6533d_row52_col0, #T_6533d_row52_col1, #T_6533d_row52_col2, #T_6533d_row53_col0, #T_6533d_row53_col1, #T_6533d_row53_col2, #T_6533d_row54_col0, #T_6533d_row54_col1, #T_6533d_row54_col2, #T_6533d_row55_col0, #T_6533d_row55_col1, #T_6533d_row55_col2, #T_6533d_row56_col0, #T_6533d_row56_col1, #T_6533d_row56_col2, #T_6533d_row57_col0, #T_6533d_row57_col1, #T_6533d_row57_col2, #T_6533d_row58_col0, #T_6533d_row58_col1, #T_6533d_row58_col2, #T_6533d_row59_col0, #T_6533d_row59_col1, #T_6533d_row59_col2, #T_6533d_row60_col0, #T_6533d_row60_col1, #T_6533d_row60_col2, #T_6533d_row61_col0, #T_6533d_row61_col1, #T_6533d_row61_col2, #T_6533d_row62_col0, #T_6533d_row62_col1, #T_6533d_row62_col2, #T_6533d_row63_col0, #T_6533d_row63_col1, #T_6533d_row63_col2, #T_6533d_row64_col0, #T_6533d_row64_col1, #T_6533d_row64_col2, #T_6533d_row65_col0, #T_6533d_row65_col1, #T_6533d_row65_col2, #T_6533d_row66_col0, #T_6533d_row66_col1, #T_6533d_row66_col2, #T_6533d_row67_col0, #T_6533d_row67_col1, #T_6533d_row67_col2, #T_6533d_row68_col0, #T_6533d_row68_col1, #T_6533d_row68_col2, #T_6533d_row69_col0, #T_6533d_row69_col1, #T_6533d_row69_col2, #T_6533d_row70_col0, #T_6533d_row70_col1, #T_6533d_row70_col2, #T_6533d_row71_col0, #T_6533d_row71_col1, #T_6533d_row71_col2, #T_6533d_row72_col0, #T_6533d_row72_col1, #T_6533d_row72_col2, #T_6533d_row73_col0, #T_6533d_row73_col1, #T_6533d_row73_col2, #T_6533d_row74_col0, #T_6533d_row74_col1, #T_6533d_row74_col2, #T_6533d_row75_col0, #T_6533d_row75_col1, #T_6533d_row75_col2, #T_6533d_row76_col0, #T_6533d_row76_col1, #T_6533d_row76_col2, #T_6533d_row77_col0, #T_6533d_row77_col1, #T_6533d_row77_col2, #T_6533d_row78_col0, #T_6533d_row78_col1, #T_6533d_row78_col2, #T_6533d_row79_col0, #T_6533d_row79_col1, #T_6533d_row79_col2, #T_6533d_row80_col0, #T_6533d_row80_col1, #T_6533d_row80_col2, #T_6533d_row81_col0, #T_6533d_row81_col1, #T_6533d_row81_col2, #T_6533d_row82_col0, #T_6533d_row82_col1, #T_6533d_row82_col2, #T_6533d_row83_col0, #T_6533d_row83_col1, #T_6533d_row83_col2, #T_6533d_row84_col0, #T_6533d_row84_col1, #T_6533d_row84_col2, #T_6533d_row85_col0, #T_6533d_row85_col1, #T_6533d_row85_col2, #T_6533d_row86_col0, #T_6533d_row86_col1, #T_6533d_row86_col2, #T_6533d_row87_col0, #T_6533d_row87_col1, #T_6533d_row87_col2, #T_6533d_row88_col0, #T_6533d_row88_col1, #T_6533d_row88_col2, #T_6533d_row89_col0, #T_6533d_row89_col1, #T_6533d_row89_col2, #T_6533d_row90_col0, #T_6533d_row90_col1, #T_6533d_row90_col2, #T_6533d_row91_col0, #T_6533d_row91_col1, #T_6533d_row91_col2, #T_6533d_row92_col0, #T_6533d_row92_col1, #T_6533d_row92_col2, #T_6533d_row93_col0, #T_6533d_row93_col1, #T_6533d_row93_col2, #T_6533d_row94_col0, #T_6533d_row94_col1, #T_6533d_row94_col2, #T_6533d_row95_col0, #T_6533d_row95_col1, #T_6533d_row95_col2, #T_6533d_row96_col0, #T_6533d_row96_col1, #T_6533d_row96_col2, #T_6533d_row97_col0, #T_6533d_row97_col1, #T_6533d_row97_col2, #T_6533d_row98_col0, #T_6533d_row98_col1, #T_6533d_row98_col2, #T_6533d_row99_col0, #T_6533d_row99_col1, #T_6533d_row99_col2, #T_6533d_row100_col0, #T_6533d_row100_col1, #T_6533d_row100_col2, #T_6533d_row101_col0, #T_6533d_row101_col1, #T_6533d_row101_col2, #T_6533d_row102_col0, #T_6533d_row102_col1, #T_6533d_row102_col2, #T_6533d_row103_col0, #T_6533d_row103_col1, #T_6533d_row103_col2, #T_6533d_row104_col0, #T_6533d_row104_col1, #T_6533d_row104_col2, #T_6533d_row105_col0, #T_6533d_row105_col1, #T_6533d_row105_col2, #T_6533d_row106_col0, #T_6533d_row106_col1, #T_6533d_row106_col2, #T_6533d_row107_col0, #T_6533d_row107_col1, #T_6533d_row107_col2, #T_6533d_row108_col0, #T_6533d_row108_col1, #T_6533d_row108_col2 {
  text-align: left;
}
</style>

<table id="T_6533d" class="caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_6533d_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">Species</th>
<th id="T_6533d_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">Latin name</th>
<th id="T_6533d_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">Assemblies</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_6533d_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</td>
<td id="T_6533d_row0_col0" class="data row0 col0">A. gambiae</td>
<td id="T_6533d_row0_col1" class="data row0 col1">Anopheles gambiae</td>
<td id="T_6533d_row0_col2" class="data row0 col2">anoGam1 , anoGam3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</td>
<td id="T_6533d_row1_col0" class="data row1 col0">A. mellifera</td>
<td id="T_6533d_row1_col1" class="data row1 col1">Apis mellifera</td>
<td id="T_6533d_row1_col2" class="data row1 col2">apiMel1 , apiMel2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">2</td>
<td id="T_6533d_row2_col0" class="data row2 col0">African clawed frog</td>
<td id="T_6533d_row2_col1" class="data row2 col1">Xenopus laevis</td>
<td id="T_6533d_row2_col2" class="data row2 col2">xenLae2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">3</td>
<td id="T_6533d_row3_col0" class="data row3 col0">Alpaca</td>
<td id="T_6533d_row3_col1" class="data row3 col1">Vicugna pacos</td>
<td id="T_6533d_row3_col2" class="data row3 col2">vicPac1 , vicPac2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">4</td>
<td id="T_6533d_row4_col0" class="data row4 col0">American alligator</td>
<td id="T_6533d_row4_col1" class="data row4 col1">Alligator mississippiensis</td>
<td id="T_6533d_row4_col2" class="data row4 col2">allMis1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row5" class="row_heading level0 row5" data-quarto-table-cell-role="th">5</td>
<td id="T_6533d_row5_col0" class="data row5 col0">Armadillo</td>
<td id="T_6533d_row5_col1" class="data row5 col1">Dasypus novemcinctus</td>
<td id="T_6533d_row5_col2" class="data row5 col2">dasNov3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row6" class="row_heading level0 row6" data-quarto-table-cell-role="th">6</td>
<td id="T_6533d_row6_col0" class="data row6 col0">Atlantic cod</td>
<td id="T_6533d_row6_col1" class="data row6 col1">Gadus morhua</td>
<td id="T_6533d_row6_col2" class="data row6 col2">gadMor1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row7" class="row_heading level0 row7" data-quarto-table-cell-role="th">7</td>
<td id="T_6533d_row7_col0" class="data row7 col0">Baboon</td>
<td id="T_6533d_row7_col1" class="data row7 col1">Papio anubis</td>
<td id="T_6533d_row7_col2" class="data row7 col2">papAnu2 , papAnu4</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row8" class="row_heading level0 row8" data-quarto-table-cell-role="th">8</td>
<td id="T_6533d_row8_col0" class="data row8 col0">Baboon</td>
<td id="T_6533d_row8_col1" class="data row8 col1">Papio hamadryas</td>
<td id="T_6533d_row8_col2" class="data row8 col2">papHam1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row9" class="row_heading level0 row9" data-quarto-table-cell-role="th">9</td>
<td id="T_6533d_row9_col0" class="data row9 col0">Bison</td>
<td id="T_6533d_row9_col1" class="data row9 col1">Bison bison bison</td>
<td id="T_6533d_row9_col2" class="data row9 col2">bisBis1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row10" class="row_heading level0 row10" data-quarto-table-cell-role="th">10</td>
<td id="T_6533d_row10_col0" class="data row10 col0">Bonobo</td>
<td id="T_6533d_row10_col1" class="data row10 col1">Pan paniscus</td>
<td id="T_6533d_row10_col2" class="data row10 col2">panPan1 , panPan2 , panPan3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row11" class="row_heading level0 row11" data-quarto-table-cell-role="th">11</td>
<td id="T_6533d_row11_col0" class="data row11 col0">Brown kiwi</td>
<td id="T_6533d_row11_col1" class="data row11 col1">Apteryx australis</td>
<td id="T_6533d_row11_col2" class="data row11 col2">aptMan1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row12" class="row_heading level0 row12" data-quarto-table-cell-role="th">12</td>
<td id="T_6533d_row12_col0" class="data row12 col0">Budgerigar</td>
<td id="T_6533d_row12_col1" class="data row12 col1">Melopsittacus undulatus</td>
<td id="T_6533d_row12_col2" class="data row12 col2">melUnd1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row13" class="row_heading level0 row13" data-quarto-table-cell-role="th">13</td>
<td id="T_6533d_row13_col0" class="data row13 col0">Bushbaby</td>
<td id="T_6533d_row13_col1" class="data row13 col1">Otolemur garnettii</td>
<td id="T_6533d_row13_col2" class="data row13 col2">otoGar3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row14" class="row_heading level0 row14" data-quarto-table-cell-role="th">14</td>
<td id="T_6533d_row14_col0" class="data row14 col0">C. brenneri</td>
<td id="T_6533d_row14_col1" class="data row14 col1">Caenorhabditis brenneri</td>
<td id="T_6533d_row14_col2" class="data row14 col2">caePb1 , caePb2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row15" class="row_heading level0 row15" data-quarto-table-cell-role="th">15</td>
<td id="T_6533d_row15_col0" class="data row15 col0">C. briggsae</td>
<td id="T_6533d_row15_col1" class="data row15 col1">Caenorhabditis briggsae</td>
<td id="T_6533d_row15_col2" class="data row15 col2">cb1 , cb3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row16" class="row_heading level0 row16" data-quarto-table-cell-role="th">16</td>
<td id="T_6533d_row16_col0" class="data row16 col0">C. elegans</td>
<td id="T_6533d_row16_col1" class="data row16 col1">Caenorhabditis elegans</td>
<td id="T_6533d_row16_col2" class="data row16 col2">ce10 , ce11 , ce2 , ce4 , ce6</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row17" class="row_heading level0 row17" data-quarto-table-cell-role="th">17</td>
<td id="T_6533d_row17_col0" class="data row17 col0">C. intestinalis</td>
<td id="T_6533d_row17_col1" class="data row17 col1">Ciona intestinalis</td>
<td id="T_6533d_row17_col2" class="data row17 col2">ci1 , ci2 , ci3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row18" class="row_heading level0 row18" data-quarto-table-cell-role="th">18</td>
<td id="T_6533d_row18_col0" class="data row18 col0">C. japonica</td>
<td id="T_6533d_row18_col1" class="data row18 col1">Caenorhabditis japonica</td>
<td id="T_6533d_row18_col2" class="data row18 col2">caeJap1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row19" class="row_heading level0 row19" data-quarto-table-cell-role="th">19</td>
<td id="T_6533d_row19_col0" class="data row19 col0">C. remanei</td>
<td id="T_6533d_row19_col1" class="data row19 col1">Caenorhabditis remanei</td>
<td id="T_6533d_row19_col2" class="data row19 col2">caeRem2 , caeRem3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row20" class="row_heading level0 row20" data-quarto-table-cell-role="th">20</td>
<td id="T_6533d_row20_col0" class="data row20 col0">Cat</td>
<td id="T_6533d_row20_col1" class="data row20 col1">Felis catus</td>
<td id="T_6533d_row20_col2" class="data row20 col2">felCat3 , felCat4 , felCat5 , felCat8 , felCat9</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row21" class="row_heading level0 row21" data-quarto-table-cell-role="th">21</td>
<td id="T_6533d_row21_col0" class="data row21 col0">Chicken</td>
<td id="T_6533d_row21_col1" class="data row21 col1">Gallus gallus</td>
<td id="T_6533d_row21_col2" class="data row21 col2">galGal2 , galGal3 , galGal4 , galGal5 , galGal6</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row22" class="row_heading level0 row22" data-quarto-table-cell-role="th">22</td>
<td id="T_6533d_row22_col0" class="data row22 col0">Chimp</td>
<td id="T_6533d_row22_col1" class="data row22 col1">Pan troglodytes</td>
<td id="T_6533d_row22_col2" class="data row22 col2">panTro1 , panTro2 , panTro3 , panTro4 , panTro5 , panTro6</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row23" class="row_heading level0 row23" data-quarto-table-cell-role="th">23</td>
<td id="T_6533d_row23_col0" class="data row23 col0">Chinese hamster</td>
<td id="T_6533d_row23_col1" class="data row23 col1">Cricetulus griseus</td>
<td id="T_6533d_row23_col2" class="data row23 col2">criGri1 , criGriChoV1, criGriChoV2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row24" class="row_heading level0 row24" data-quarto-table-cell-role="th">24</td>
<td id="T_6533d_row24_col0" class="data row24 col0">Chinese pangolin</td>
<td id="T_6533d_row24_col1" class="data row24 col1">Manis pentadactyla</td>
<td id="T_6533d_row24_col2" class="data row24 col2">manPen1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row25" class="row_heading level0 row25" data-quarto-table-cell-role="th">25</td>
<td id="T_6533d_row25_col0" class="data row25 col0">Coelacanth</td>
<td id="T_6533d_row25_col1" class="data row25 col1">Latimeria chalumnae</td>
<td id="T_6533d_row25_col2" class="data row25 col2">latCha1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row26" class="row_heading level0 row26" data-quarto-table-cell-role="th">26</td>
<td id="T_6533d_row26_col0" class="data row26 col0">Cow</td>
<td id="T_6533d_row26_col1" class="data row26 col1">Bos taurus</td>
<td id="T_6533d_row26_col2" class="data row26 col2">bosTau2 , bosTau3 , bosTau4 , bosTau6 , bosTau7 , bosTau8 , bosTau9</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row27" class="row_heading level0 row27" data-quarto-table-cell-role="th">27</td>
<td id="T_6533d_row27_col0" class="data row27 col0">Crab-eating macaque</td>
<td id="T_6533d_row27_col1" class="data row27 col1">Macaca fascicularis</td>
<td id="T_6533d_row27_col2" class="data row27 col2">macFas5</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row28" class="row_heading level0 row28" data-quarto-table-cell-role="th">28</td>
<td id="T_6533d_row28_col0" class="data row28 col0">D. ananassae</td>
<td id="T_6533d_row28_col1" class="data row28 col1">Drosophila ananassae</td>
<td id="T_6533d_row28_col2" class="data row28 col2">droAna1 , droAna2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row29" class="row_heading level0 row29" data-quarto-table-cell-role="th">29</td>
<td id="T_6533d_row29_col0" class="data row29 col0">D. erecta</td>
<td id="T_6533d_row29_col1" class="data row29 col1">Drosophila erecta</td>
<td id="T_6533d_row29_col2" class="data row29 col2">droEre1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row30" class="row_heading level0 row30" data-quarto-table-cell-role="th">30</td>
<td id="T_6533d_row30_col0" class="data row30 col0">D. grimshawi</td>
<td id="T_6533d_row30_col1" class="data row30 col1">Drosophila grimshawi</td>
<td id="T_6533d_row30_col2" class="data row30 col2">droGri1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row31" class="row_heading level0 row31" data-quarto-table-cell-role="th">31</td>
<td id="T_6533d_row31_col0" class="data row31 col0">D. melanogaster</td>
<td id="T_6533d_row31_col1" class="data row31 col1">Drosophila melanogaster</td>
<td id="T_6533d_row31_col2" class="data row31 col2">dm1 , dm2 , dm3 , dm6</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row32" class="row_heading level0 row32" data-quarto-table-cell-role="th">32</td>
<td id="T_6533d_row32_col0" class="data row32 col0">D. mojavensis</td>
<td id="T_6533d_row32_col1" class="data row32 col1">Drosophila mojavensis</td>
<td id="T_6533d_row32_col2" class="data row32 col2">droMoj1 , droMoj2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row33" class="row_heading level0 row33" data-quarto-table-cell-role="th">33</td>
<td id="T_6533d_row33_col0" class="data row33 col0">D. persimilis</td>
<td id="T_6533d_row33_col1" class="data row33 col1">Drosophila persimilis</td>
<td id="T_6533d_row33_col2" class="data row33 col2">droPer1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row34" class="row_heading level0 row34" data-quarto-table-cell-role="th">34</td>
<td id="T_6533d_row34_col0" class="data row34 col0">D. pseudoobscura</td>
<td id="T_6533d_row34_col1" class="data row34 col1">Drosophila pseudoobscura</td>
<td id="T_6533d_row34_col2" class="data row34 col2">dp2 , dp3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row35" class="row_heading level0 row35" data-quarto-table-cell-role="th">35</td>
<td id="T_6533d_row35_col0" class="data row35 col0">D. sechellia</td>
<td id="T_6533d_row35_col1" class="data row35 col1">Drosophila sechellia</td>
<td id="T_6533d_row35_col2" class="data row35 col2">droSec1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row36" class="row_heading level0 row36" data-quarto-table-cell-role="th">36</td>
<td id="T_6533d_row36_col0" class="data row36 col0">D. simulans</td>
<td id="T_6533d_row36_col1" class="data row36 col1">Drosophila simulans</td>
<td id="T_6533d_row36_col2" class="data row36 col2">droSim1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row37" class="row_heading level0 row37" data-quarto-table-cell-role="th">37</td>
<td id="T_6533d_row37_col0" class="data row37 col0">D. virilis</td>
<td id="T_6533d_row37_col1" class="data row37 col1">Drosophila virilis</td>
<td id="T_6533d_row37_col2" class="data row37 col2">droVir1 , droVir2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row38" class="row_heading level0 row38" data-quarto-table-cell-role="th">38</td>
<td id="T_6533d_row38_col0" class="data row38 col0">D. yakuba</td>
<td id="T_6533d_row38_col1" class="data row38 col1">Drosophila yakuba</td>
<td id="T_6533d_row38_col2" class="data row38 col2">droYak1 , droYak2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row39" class="row_heading level0 row39" data-quarto-table-cell-role="th">39</td>
<td id="T_6533d_row39_col0" class="data row39 col0">Dog</td>
<td id="T_6533d_row39_col1" class="data row39 col1">Canis lupus familiaris</td>
<td id="T_6533d_row39_col2" class="data row39 col2">canFam1 , canFam2 , canFam3 , canFam4 , canFam5 , canFam6</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row40" class="row_heading level0 row40" data-quarto-table-cell-role="th">40</td>
<td id="T_6533d_row40_col0" class="data row40 col0">Dolphin</td>
<td id="T_6533d_row40_col1" class="data row40 col1">Tursiops truncatus</td>
<td id="T_6533d_row40_col2" class="data row40 col2">turTru2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row41" class="row_heading level0 row41" data-quarto-table-cell-role="th">41</td>
<td id="T_6533d_row41_col0" class="data row41 col0">Ebola virus</td>
<td id="T_6533d_row41_col1" class="data row41 col1">Filoviridae ebolavirus</td>
<td id="T_6533d_row41_col2" class="data row41 col2">eboVir3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row42" class="row_heading level0 row42" data-quarto-table-cell-role="th">42</td>
<td id="T_6533d_row42_col0" class="data row42 col0">Elephant</td>
<td id="T_6533d_row42_col1" class="data row42 col1">Loxodonta africana</td>
<td id="T_6533d_row42_col2" class="data row42 col2">loxAfr3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row43" class="row_heading level0 row43" data-quarto-table-cell-role="th">43</td>
<td id="T_6533d_row43_col0" class="data row43 col0">Elephant shark</td>
<td id="T_6533d_row43_col1" class="data row43 col1">Callorhinchus milii</td>
<td id="T_6533d_row43_col2" class="data row43 col2">calMil1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row44" class="row_heading level0 row44" data-quarto-table-cell-role="th">44</td>
<td id="T_6533d_row44_col0" class="data row44 col0">Ferret</td>
<td id="T_6533d_row44_col1" class="data row44 col1">Mustela putorius furo</td>
<td id="T_6533d_row44_col2" class="data row44 col2">musFur1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row45" class="row_heading level0 row45" data-quarto-table-cell-role="th">45</td>
<td id="T_6533d_row45_col0" class="data row45 col0">Fugu</td>
<td id="T_6533d_row45_col1" class="data row45 col1">Takifugu rubripes</td>
<td id="T_6533d_row45_col2" class="data row45 col2">fr1 , fr2 , fr3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row46" class="row_heading level0 row46" data-quarto-table-cell-role="th">46</td>
<td id="T_6533d_row46_col0" class="data row46 col0">Garter snake</td>
<td id="T_6533d_row46_col1" class="data row46 col1">Thamnophis sirtalis</td>
<td id="T_6533d_row46_col2" class="data row46 col2">thaSir1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row47" class="row_heading level0 row47" data-quarto-table-cell-role="th">47</td>
<td id="T_6533d_row47_col0" class="data row47 col0">Gibbon</td>
<td id="T_6533d_row47_col1" class="data row47 col1">Nomascus leucogenys</td>
<td id="T_6533d_row47_col2" class="data row47 col2">nomLeu1 , nomLeu2 , nomLeu3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row48" class="row_heading level0 row48" data-quarto-table-cell-role="th">48</td>
<td id="T_6533d_row48_col0" class="data row48 col0">Golden eagle</td>
<td id="T_6533d_row48_col1" class="data row48 col1">Aquila chrysaetos canadensis</td>
<td id="T_6533d_row48_col2" class="data row48 col2">aquChr2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row49" class="row_heading level0 row49" data-quarto-table-cell-role="th">49</td>
<td id="T_6533d_row49_col0" class="data row49 col0">Golden snub-nosed monkey</td>
<td id="T_6533d_row49_col1" class="data row49 col1">Rhinopithecus roxellana</td>
<td id="T_6533d_row49_col2" class="data row49 col2">rhiRox1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row50" class="row_heading level0 row50" data-quarto-table-cell-role="th">50</td>
<td id="T_6533d_row50_col0" class="data row50 col0">Gorilla</td>
<td id="T_6533d_row50_col1" class="data row50 col1">Gorilla gorilla gorilla</td>
<td id="T_6533d_row50_col2" class="data row50 col2">gorGor3 , gorGor4 , gorGor5 , gorGor6</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row51" class="row_heading level0 row51" data-quarto-table-cell-role="th">51</td>
<td id="T_6533d_row51_col0" class="data row51 col0">Green monkey</td>
<td id="T_6533d_row51_col1" class="data row51 col1">Chlorocebus sabaeus</td>
<td id="T_6533d_row51_col2" class="data row51 col2">chlSab2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row52" class="row_heading level0 row52" data-quarto-table-cell-role="th">52</td>
<td id="T_6533d_row52_col0" class="data row52 col0">Guinea pig</td>
<td id="T_6533d_row52_col1" class="data row52 col1">Cavia porcellus</td>
<td id="T_6533d_row52_col2" class="data row52 col2">cavPor3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row53" class="row_heading level0 row53" data-quarto-table-cell-role="th">53</td>
<td id="T_6533d_row53_col0" class="data row53 col0">Hawaiian monk seal</td>
<td id="T_6533d_row53_col1" class="data row53 col1">Neomonachus schauinslandi</td>
<td id="T_6533d_row53_col2" class="data row53 col2">neoSch1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row54" class="row_heading level0 row54" data-quarto-table-cell-role="th">54</td>
<td id="T_6533d_row54_col0" class="data row54 col0">Hedgehog</td>
<td id="T_6533d_row54_col1" class="data row54 col1">Erinaceus europaeus</td>
<td id="T_6533d_row54_col2" class="data row54 col2">eriEur1 , eriEur2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row55" class="row_heading level0 row55" data-quarto-table-cell-role="th">55</td>
<td id="T_6533d_row55_col0" class="data row55 col0">Horse</td>
<td id="T_6533d_row55_col1" class="data row55 col1">Equus caballus</td>
<td id="T_6533d_row55_col2" class="data row55 col2">equCab1 , equCab2 , equCab3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row56" class="row_heading level0 row56" data-quarto-table-cell-role="th">56</td>
<td id="T_6533d_row56_col0" class="data row56 col0">Human</td>
<td id="T_6533d_row56_col1" class="data row56 col1">Homo sapiens</td>
<td id="T_6533d_row56_col2" class="data row56 col2">hg16 , hg17 , hg18 , hg19 , hg38 , hs1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row57" class="row_heading level0 row57" data-quarto-table-cell-role="th">57</td>
<td id="T_6533d_row57_col0" class="data row57 col0">Kangaroo rat</td>
<td id="T_6533d_row57_col1" class="data row57 col1">Dipodomys ordii</td>
<td id="T_6533d_row57_col2" class="data row57 col2">dipOrd1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row58" class="row_heading level0 row58" data-quarto-table-cell-role="th">58</td>
<td id="T_6533d_row58_col0" class="data row58 col0">Lamprey</td>
<td id="T_6533d_row58_col1" class="data row58 col1">Petromyzon marinus</td>
<td id="T_6533d_row58_col2" class="data row58 col2">petMar1 , petMar2 , petMar3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row59" class="row_heading level0 row59" data-quarto-table-cell-role="th">59</td>
<td id="T_6533d_row59_col0" class="data row59 col0">Lancelet</td>
<td id="T_6533d_row59_col1" class="data row59 col1">Branchiostoma floridae</td>
<td id="T_6533d_row59_col2" class="data row59 col2">braFlo1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row60" class="row_heading level0 row60" data-quarto-table-cell-role="th">60</td>
<td id="T_6533d_row60_col0" class="data row60 col0">Lizard</td>
<td id="T_6533d_row60_col1" class="data row60 col1">Anolis carolinensis</td>
<td id="T_6533d_row60_col2" class="data row60 col2">anoCar1 , anoCar2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row61" class="row_heading level0 row61" data-quarto-table-cell-role="th">61</td>
<td id="T_6533d_row61_col0" class="data row61 col0">Malayan flying lemur</td>
<td id="T_6533d_row61_col1" class="data row61 col1">Galeopterus variegatus</td>
<td id="T_6533d_row61_col2" class="data row61 col2">galVar1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row62" class="row_heading level0 row62" data-quarto-table-cell-role="th">62</td>
<td id="T_6533d_row62_col0" class="data row62 col0">Manatee</td>
<td id="T_6533d_row62_col1" class="data row62 col1">Trichechus manatus latirostris</td>
<td id="T_6533d_row62_col2" class="data row62 col2">triMan1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row63" class="row_heading level0 row63" data-quarto-table-cell-role="th">63</td>
<td id="T_6533d_row63_col0" class="data row63 col0">Marmoset</td>
<td id="T_6533d_row63_col1" class="data row63 col1">Callithrix jacchus</td>
<td id="T_6533d_row63_col2" class="data row63 col2">calJac1 , calJac3 , calJac4</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row64" class="row_heading level0 row64" data-quarto-table-cell-role="th">64</td>
<td id="T_6533d_row64_col0" class="data row64 col0">Medaka</td>
<td id="T_6533d_row64_col1" class="data row64 col1">Oryzias latipes</td>
<td id="T_6533d_row64_col2" class="data row64 col2">oryLat2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row65" class="row_heading level0 row65" data-quarto-table-cell-role="th">65</td>
<td id="T_6533d_row65_col0" class="data row65 col0">Medium ground finch</td>
<td id="T_6533d_row65_col1" class="data row65 col1">Geospiza fortis</td>
<td id="T_6533d_row65_col2" class="data row65 col2">geoFor1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row66" class="row_heading level0 row66" data-quarto-table-cell-role="th">66</td>
<td id="T_6533d_row66_col0" class="data row66 col0">Megabat</td>
<td id="T_6533d_row66_col1" class="data row66 col1">Pteropus vampyrus</td>
<td id="T_6533d_row66_col2" class="data row66 col2">pteVam1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row67" class="row_heading level0 row67" data-quarto-table-cell-role="th">67</td>
<td id="T_6533d_row67_col0" class="data row67 col0">Minke whale</td>
<td id="T_6533d_row67_col1" class="data row67 col1">Balaenoptera acutorostrata scammoni</td>
<td id="T_6533d_row67_col2" class="data row67 col2">balAcu1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row68" class="row_heading level0 row68" data-quarto-table-cell-role="th">68</td>
<td id="T_6533d_row68_col0" class="data row68 col0">Monkeypox virus</td>
<td id="T_6533d_row68_col1" class="data row68 col1">Monkeypox virus</td>
<td id="T_6533d_row68_col2" class="data row68 col2">mpxvRivers</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row69" class="row_heading level0 row69" data-quarto-table-cell-role="th">69</td>
<td id="T_6533d_row69_col0" class="data row69 col0">Mouse</td>
<td id="T_6533d_row69_col1" class="data row69 col1">Mus musculus</td>
<td id="T_6533d_row69_col2" class="data row69 col2">mm10 , mm39 , mm7 , mm8 , mm9</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row70" class="row_heading level0 row70" data-quarto-table-cell-role="th">70</td>
<td id="T_6533d_row70_col0" class="data row70 col0">Mouse lemur</td>
<td id="T_6533d_row70_col1" class="data row70 col1">Microcebus murinus</td>
<td id="T_6533d_row70_col2" class="data row70 col2">micMur1 , micMur2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row71" class="row_heading level0 row71" data-quarto-table-cell-role="th">71</td>
<td id="T_6533d_row71_col0" class="data row71 col0">Naked mole-rat</td>
<td id="T_6533d_row71_col1" class="data row71 col1">Heterocephalus glaber</td>
<td id="T_6533d_row71_col2" class="data row71 col2">hetGla1 , hetGla2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row72" class="row_heading level0 row72" data-quarto-table-cell-role="th">72</td>
<td id="T_6533d_row72_col0" class="data row72 col0">Nile tilapia</td>
<td id="T_6533d_row72_col1" class="data row72 col1">Oreochromis niloticus</td>
<td id="T_6533d_row72_col2" class="data row72 col2">oreNil2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row73" class="row_heading level0 row73" data-quarto-table-cell-role="th">73</td>
<td id="T_6533d_row73_col0" class="data row73 col0">Opossum</td>
<td id="T_6533d_row73_col1" class="data row73 col1">Monodelphis domestica</td>
<td id="T_6533d_row73_col2" class="data row73 col2">monDom1 , monDom4 , monDom5</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row74" class="row_heading level0 row74" data-quarto-table-cell-role="th">74</td>
<td id="T_6533d_row74_col0" class="data row74 col0">Orangutan</td>
<td id="T_6533d_row74_col1" class="data row74 col1">Pongo pygmaeus abelii</td>
<td id="T_6533d_row74_col2" class="data row74 col2">ponAbe2 , ponAbe3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row75" class="row_heading level0 row75" data-quarto-table-cell-role="th">75</td>
<td id="T_6533d_row75_col0" class="data row75 col0">P. pacificus</td>
<td id="T_6533d_row75_col1" class="data row75 col1">Pristionchus pacificus</td>
<td id="T_6533d_row75_col2" class="data row75 col2">priPac1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row76" class="row_heading level0 row76" data-quarto-table-cell-role="th">76</td>
<td id="T_6533d_row76_col0" class="data row76 col0">Painted turtle</td>
<td id="T_6533d_row76_col1" class="data row76 col1">Chrysemys picta bellii</td>
<td id="T_6533d_row76_col2" class="data row76 col2">chrPic1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row77" class="row_heading level0 row77" data-quarto-table-cell-role="th">77</td>
<td id="T_6533d_row77_col0" class="data row77 col0">Panda</td>
<td id="T_6533d_row77_col1" class="data row77 col1">Ailuropoda melanoleuca</td>
<td id="T_6533d_row77_col2" class="data row77 col2">ailMel1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row78" class="row_heading level0 row78" data-quarto-table-cell-role="th">78</td>
<td id="T_6533d_row78_col0" class="data row78 col0">Pig</td>
<td id="T_6533d_row78_col1" class="data row78 col1">Sus scrofa</td>
<td id="T_6533d_row78_col2" class="data row78 col2">susScr11, susScr2 , susScr3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row79" class="row_heading level0 row79" data-quarto-table-cell-role="th">79</td>
<td id="T_6533d_row79_col0" class="data row79 col0">Pika</td>
<td id="T_6533d_row79_col1" class="data row79 col1">Ochotona princeps</td>
<td id="T_6533d_row79_col2" class="data row79 col2">ochPri2 , ochPri3</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row80" class="row_heading level0 row80" data-quarto-table-cell-role="th">80</td>
<td id="T_6533d_row80_col0" class="data row80 col0">Platypus</td>
<td id="T_6533d_row80_col1" class="data row80 col1">Ornithorhynchus anatinus</td>
<td id="T_6533d_row80_col2" class="data row80 col2">ornAna1 , ornAna2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row81" class="row_heading level0 row81" data-quarto-table-cell-role="th">81</td>
<td id="T_6533d_row81_col0" class="data row81 col0">Proboscis monkey</td>
<td id="T_6533d_row81_col1" class="data row81 col1">Nasalis larvatus</td>
<td id="T_6533d_row81_col2" class="data row81 col2">nasLar1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row82" class="row_heading level0 row82" data-quarto-table-cell-role="th">82</td>
<td id="T_6533d_row82_col0" class="data row82 col0">Rabbit</td>
<td id="T_6533d_row82_col1" class="data row82 col1">Oryctolagus cuniculus</td>
<td id="T_6533d_row82_col2" class="data row82 col2">oryCun2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row83" class="row_heading level0 row83" data-quarto-table-cell-role="th">83</td>
<td id="T_6533d_row83_col0" class="data row83 col0">Rat</td>
<td id="T_6533d_row83_col1" class="data row83 col1">Rattus norvegicus</td>
<td id="T_6533d_row83_col2" class="data row83 col2">rn3 , rn4 , rn5 , rn6 , rn7</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row84" class="row_heading level0 row84" data-quarto-table-cell-role="th">84</td>
<td id="T_6533d_row84_col0" class="data row84 col0">Rhesus</td>
<td id="T_6533d_row84_col1" class="data row84 col1">Macaca mulatta</td>
<td id="T_6533d_row84_col2" class="data row84 col2">rheMac10, rheMac2 , rheMac3 , rheMac8</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row85" class="row_heading level0 row85" data-quarto-table-cell-role="th">85</td>
<td id="T_6533d_row85_col0" class="data row85 col0">Rock hyrax</td>
<td id="T_6533d_row85_col1" class="data row85 col1">Procavia capensis</td>
<td id="T_6533d_row85_col2" class="data row85 col2">proCap1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row86" class="row_heading level0 row86" data-quarto-table-cell-role="th">86</td>
<td id="T_6533d_row86_col0" class="data row86 col0">S. cerevisiae</td>
<td id="T_6533d_row86_col1" class="data row86 col1">Saccharomyces cerevisiae</td>
<td id="T_6533d_row86_col2" class="data row86 col2">sacCer1 , sacCer2 , sacCer3</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row87" class="row_heading level0 row87" data-quarto-table-cell-role="th">87</td>
<td id="T_6533d_row87_col0" class="data row87 col0">S. purpuratus</td>
<td id="T_6533d_row87_col1" class="data row87 col1">Strongylocentrotus purpuratus</td>
<td id="T_6533d_row87_col2" class="data row87 col2">strPur1 , strPur2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row88" class="row_heading level0 row88" data-quarto-table-cell-role="th">88</td>
<td id="T_6533d_row88_col0" class="data row88 col0">SARS-CoV-2</td>
<td id="T_6533d_row88_col1" class="data row88 col1">SARS-CoV-2</td>
<td id="T_6533d_row88_col2" class="data row88 col2">wuhCor1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row89" class="row_heading level0 row89" data-quarto-table-cell-role="th">89</td>
<td id="T_6533d_row89_col0" class="data row89 col0">Sea hare</td>
<td id="T_6533d_row89_col1" class="data row89 col1">Aplysia californica</td>
<td id="T_6533d_row89_col2" class="data row89 col2">aplCal1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row90" class="row_heading level0 row90" data-quarto-table-cell-role="th">90</td>
<td id="T_6533d_row90_col0" class="data row90 col0">Sheep</td>
<td id="T_6533d_row90_col1" class="data row90 col1">Ovis aries</td>
<td id="T_6533d_row90_col2" class="data row90 col2">oviAri1 , oviAri3 , oviAri4</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row91" class="row_heading level0 row91" data-quarto-table-cell-role="th">91</td>
<td id="T_6533d_row91_col0" class="data row91 col0">Shrew</td>
<td id="T_6533d_row91_col1" class="data row91 col1">Sorex araneus</td>
<td id="T_6533d_row91_col2" class="data row91 col2">sorAra1 , sorAra2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row92" class="row_heading level0 row92" data-quarto-table-cell-role="th">92</td>
<td id="T_6533d_row92_col0" class="data row92 col0">Sloth</td>
<td id="T_6533d_row92_col1" class="data row92 col1">Choloepus hoffmanni</td>
<td id="T_6533d_row92_col2" class="data row92 col2">choHof1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row93" class="row_heading level0 row93" data-quarto-table-cell-role="th">93</td>
<td id="T_6533d_row93_col0" class="data row93 col0">Southern sea otter</td>
<td id="T_6533d_row93_col1" class="data row93 col1">Enhydra lutris nereis</td>
<td id="T_6533d_row93_col2" class="data row93 col2">enhLutNer1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row94" class="row_heading level0 row94" data-quarto-table-cell-role="th">94</td>
<td id="T_6533d_row94_col0" class="data row94 col0">Squirrel</td>
<td id="T_6533d_row94_col1" class="data row94 col1">Spermophilus tridecemlineatus</td>
<td id="T_6533d_row94_col2" class="data row94 col2">speTri2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row95" class="row_heading level0 row95" data-quarto-table-cell-role="th">95</td>
<td id="T_6533d_row95_col0" class="data row95 col0">Squirrel monkey</td>
<td id="T_6533d_row95_col1" class="data row95 col1">Saimiri boliviensis</td>
<td id="T_6533d_row95_col2" class="data row95 col2">saiBol1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row96" class="row_heading level0 row96" data-quarto-table-cell-role="th">96</td>
<td id="T_6533d_row96_col0" class="data row96 col0">Stickleback</td>
<td id="T_6533d_row96_col1" class="data row96 col1">Gasterosteus aculeatus</td>
<td id="T_6533d_row96_col2" class="data row96 col2">gasAcu1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row97" class="row_heading level0 row97" data-quarto-table-cell-role="th">97</td>
<td id="T_6533d_row97_col0" class="data row97 col0">Tarsier</td>
<td id="T_6533d_row97_col1" class="data row97 col1">Tarsius syrichta</td>
<td id="T_6533d_row97_col2" class="data row97 col2">tarSyr1 , tarSyr2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row98" class="row_heading level0 row98" data-quarto-table-cell-role="th">98</td>
<td id="T_6533d_row98_col0" class="data row98 col0">Tasmanian devil</td>
<td id="T_6533d_row98_col1" class="data row98 col1">Sarcophilus harrisii</td>
<td id="T_6533d_row98_col2" class="data row98 col2">sarHar1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row99" class="row_heading level0 row99" data-quarto-table-cell-role="th">99</td>
<td id="T_6533d_row99_col0" class="data row99 col0">Tenrec</td>
<td id="T_6533d_row99_col1" class="data row99 col1">Echinops telfairi</td>
<td id="T_6533d_row99_col2" class="data row99 col2">echTel1 , echTel2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row100" class="row_heading level0 row100" data-quarto-table-cell-role="th">100</td>
<td id="T_6533d_row100_col0" class="data row100 col0">Tetraodon</td>
<td id="T_6533d_row100_col1" class="data row100 col1">Tetraodon nigroviridis</td>
<td id="T_6533d_row100_col2" class="data row100 col2">tetNig1 , tetNig2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row101" class="row_heading level0 row101" data-quarto-table-cell-role="th">101</td>
<td id="T_6533d_row101_col0" class="data row101 col0">Tibetan frog</td>
<td id="T_6533d_row101_col1" class="data row101 col1">Nanorana parkeri</td>
<td id="T_6533d_row101_col2" class="data row101 col2">nanPar1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row102" class="row_heading level0 row102" data-quarto-table-cell-role="th">102</td>
<td id="T_6533d_row102_col0" class="data row102 col0">Tree shrew</td>
<td id="T_6533d_row102_col1" class="data row102 col1">Tupaia belangeri</td>
<td id="T_6533d_row102_col2" class="data row102 col2">tupBel1</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row103" class="row_heading level0 row103" data-quarto-table-cell-role="th">103</td>
<td id="T_6533d_row103_col0" class="data row103 col0">Turkey</td>
<td id="T_6533d_row103_col1" class="data row103 col1">Meleagris gallopavo</td>
<td id="T_6533d_row103_col2" class="data row103 col2">melGal1 , melGal5</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row104" class="row_heading level0 row104" data-quarto-table-cell-role="th">104</td>
<td id="T_6533d_row104_col0" class="data row104 col0">Wallaby</td>
<td id="T_6533d_row104_col1" class="data row104 col1">Macropus eugenii</td>
<td id="T_6533d_row104_col2" class="data row104 col2">macEug2</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row105" class="row_heading level0 row105" data-quarto-table-cell-role="th">105</td>
<td id="T_6533d_row105_col0" class="data row105 col0">White rhinoceros</td>
<td id="T_6533d_row105_col1" class="data row105 col1">Ceratotherium simum</td>
<td id="T_6533d_row105_col2" class="data row105 col2">cerSim1</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row106" class="row_heading level0 row106" data-quarto-table-cell-role="th">106</td>
<td id="T_6533d_row106_col0" class="data row106 col0">X. tropicalis</td>
<td id="T_6533d_row106_col1" class="data row106 col1">Xenopus tropicalis</td>
<td id="T_6533d_row106_col2" class="data row106 col2">xenTro1 , xenTro10, xenTro2 , xenTro3 , xenTro7 , xenTro9</td>
</tr>
<tr class="even">
<td id="T_6533d_level0_row107" class="row_heading level0 row107" data-quarto-table-cell-role="th">107</td>
<td id="T_6533d_row107_col0" class="data row107 col0">Zebra finch</td>
<td id="T_6533d_row107_col1" class="data row107 col1">Taeniopygia guttata</td>
<td id="T_6533d_row107_col2" class="data row107 col2">taeGut1 , taeGut2</td>
</tr>
<tr class="odd">
<td id="T_6533d_level0_row108" class="row_heading level0 row108" data-quarto-table-cell-role="th">108</td>
<td id="T_6533d_row108_col0" class="data row108 col0">Zebrafish</td>
<td id="T_6533d_row108_col1" class="data row108 col1">Danio rerio</td>
<td id="T_6533d_row108_col2" class="data row108 col2">danRer10, danRer11, danRer3 , danRer4 , danRer5 , danRer6 , danRer7</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-12" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:19.878465Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:19.878299Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:23.678133Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:23.677770Z&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'panTro4'</span>) </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-10-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="chrom_ideograms_files/figure-html/cell-10-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="single-chromosome-ideogram" class="level2">
<h2 class="anchored" data-anchor-id="single-chromosome-ideogram">Single chromosome ideogram</h2>
<div id="cell-14" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:23.681134Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:23.680966Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:23.781276Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:23.780939Z&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg19'</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-11-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="chrom_ideograms_files/figure-html/cell-11-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="the-coordinate-system" class="level2">
<h2 class="anchored" data-anchor-id="the-coordinate-system">The coordinate system</h2>
<p>The x-axis follow the chromosome coordinates along the ideogram. Behind the scenes, the y-axis for each ideogram has the same scale as its x-axis. But this is mapped to a more sensible range that is <code>(0, 10)</code> by default can be controlled by the <code>ylim</code> keyword argument:</p>
<div id="cell-17" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:23.783138Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:23.783007Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:23.879280Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:23.878998Z&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, ylim<span class="op">=</span>(<span class="dv">0</span>, <span class="dv">30</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-12-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="chrom_ideograms_files/figure-html/cell-12-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ideogram-size-and-plament" class="level2">
<h2 class="anchored" data-anchor-id="ideogram-size-and-plament">Ideogram size and plament</h2>
<div id="cell-19" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:23.880858Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:23.880753Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:23.977376Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:23.977047Z&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">1</span>, height<span class="op">=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-13-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="chrom_ideograms_files/figure-html/cell-13-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="genomic-segment-annotation" class="level2">
<h2 class="anchored" data-anchor-id="genomic-segment-annotation">Genomic segment annotation</h2>
<div id="cell-21" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:23.979231Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:23.979103Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:25.416282Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:25.415944Z&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>g.add_segments(segments, facecolor<span class="op">=</span><span class="st">'tab:blue'</span>, alpha<span class="op">=</span><span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-14-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="chrom_ideograms_files/figure-html/cell-14-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>To make sure segments are visible on the plot, the default minimal size of plotteed segments are 200000. This can be changed like this (notice how the chr1 segment is no longer visible):</p>
<div id="cell-23" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:25.418726Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:25.418587Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:25.434686Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:25.434425Z&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>segments_with_alpha <span class="op">=</span> [(<span class="op">*</span>t, np.random.random()) <span class="cf">for</span> t <span class="kw">in</span> segments]</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>segments_with_alpha[:<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>[('chr1', 10065165, 12961789, 0.061182546920045655),
 ('chr1', 15568967, 18751672, 0.021441536887357615),
 ('chr1', 19444467, 23071285, 0.37882307161321216)]</code></pre>
</div>
</div>
<div id="cell-24" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:25.436154Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:25.436064Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:26.845163Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:26.844868Z&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>g.add_segments(segments_with_alpha, facecolor<span class="op">=</span><span class="st">'tab:blue'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-16-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="chrom_ideograms_files/figure-html/cell-16-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="conditional-ideogram-color" class="level2">
<h2 class="anchored" data-anchor-id="conditional-ideogram-color">Conditional ideogram color</h2>
<p>You can also make consequtive segments with individual alpha values to evalutate a statistic across the chromosomes. Below I make 100kb windows with sin values:</p>
<div id="cell-27" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:26.847288Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:26.847173Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:26.862987Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:26.862710Z&quot;}" data-execution_count="16">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>length <span class="op">=</span> utils.chrom_lengths[<span class="st">'hg38'</span>][<span class="st">'chrX'</span>]</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>win <span class="op">=</span> <span class="dv">100_000</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> <span class="bu">range</span>(<span class="dv">0</span>, length, win)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> np.linspace(<span class="dv">5</span>, <span class="dv">10</span>, <span class="bu">len</span>(x))</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> np.sin(<span class="dv">2</span><span class="op">*</span> np.pi <span class="op">*</span> f <span class="op">*</span> x <span class="op">/</span> length) <span class="op">/</span> <span class="dv">2</span> <span class="op">+</span> <span class="dv">1</span><span class="op">/</span><span class="dv">2</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>chrom <span class="op">=</span> [<span class="st">'chrX'</span>]<span class="op">*</span><span class="bu">len</span>(x)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>shaded_segments <span class="op">=</span> [(<span class="st">'chrX'</span>, s, s<span class="op">+</span>win, a) <span class="cf">for</span> s, a <span class="kw">in</span> <span class="bu">zip</span>(x, y)]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>shaded_segments[:<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>[('chrX', 0, 100000, 0.5),
 ('chrX', 100000, 200000, 0.5100723398623864),
 ('chrX', 200000, 300000, 0.5201534872481727)]</code></pre>
</div>
</div>
<div id="cell-28" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:26.864589Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:26.864475Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:27.662625Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:27.662022Z&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>g.add_segments(shaded_segments, facecolor<span class="op">=</span><span class="st">'tab:red'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-18-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="chrom_ideograms_files/figure-html/cell-18-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Even better, you can use the <code>add_horizon</code> method to plot a horizon plot on the ideograms:</p>
<div id="cell-30" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:27.667196Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:27.667077Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:29.714862Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:29.714505Z&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">3</span>, height<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>g.add_horizon(df, zorder<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-19-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="chrom_ideograms_files/figure-html/cell-19-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="context-plots" class="level2">
<h2 class="anchored" data-anchor-id="context-plots">Context plots</h2>
<p>To visualize data in chromosome context, you can plot anything you like above, on top, or below the ideogram:</p>
<div id="cell-33" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:29.803157Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:29.803014Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:31.604466Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:31.604021Z&quot;}" data-execution_count="19">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>g.map_fun(sns.scatterplot, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, yaxis<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">2</span>), s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'C1'</span>, label<span class="op">=</span><span class="st">'sin curve scatter'</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>g.add_legend()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-20-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="chrom_ideograms_files/figure-html/cell-20-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-34" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:31.620174Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:31.620033Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:33.988471Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:33.988170Z&quot;}" data-execution_count="20">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>g.map_fun(sns.lineplot, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, yaxis<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">7</span>), linewidth<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">'sin curve line'</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>g.add_legend()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-21-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="chrom_ideograms_files/figure-html/cell-21-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="labelled-positions" class="level2">
<h2 class="anchored" data-anchor-id="labelled-positions">Labelled positions</h2>
<p>You can label chromosome positions, to show gene locations or anything else:</p>
<div id="cell-37" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:33.990310Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:33.990197Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:34.005304Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:34.005072Z&quot;}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sample_genes <span class="op">=</span> gene_lists.get(<span class="st">'neuron_genome_proteome'</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sample_genes[:<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<table class="caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<tbody>
<tr class="odd" style="background: transparent!important; line-height: 10px!important;text-align: left!important">
<td style="background: transparent!important; line-height: 10px!important">A1BG</td>
<td style="background: transparent!important; line-height: 10px!important">A1CF</td>
<td style="background: transparent!important; line-height: 10px!important">A2M</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-38" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:34.006881Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:34.006768Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:25:34.018843Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:25:34.018413Z&quot;}" data-execution_count="22">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo.information <span class="im">as</span> gi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-39" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:25:34.020838Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:25:34.020720Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:10.790269Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:10.789965Z&quot;}" data-execution_count="23">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sample_gene_coord <span class="op">=</span> gi.gene_labels(random.sample(sample_genes, <span class="dv">100</span>), assembly<span class="op">=</span><span class="st">'hg38'</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>sample_gene_coord[:<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>C5orf51 not found</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>[('chr5', 181199559.5, 'TRIM7'),
 ('chr10', 1094378.0, 'WDR37'),
 ('chr6', 12263854.5, 'EDN1')]</code></pre>
</div>
</div>
<p>Add gene labels to positions:</p>
<div id="cell-41" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:10.793714Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:10.793585Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:12.519605Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:12.519315Z&quot;}" data-execution_count="24">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-25-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="chrom_ideograms_files/figure-html/cell-25-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>You can also use <code>add_labels</code> to add position markers along a chromosome like this to better see the location</p>
<div id="cell-43" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:12.522031Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:12.521901Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:16.700464Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:16.700099Z&quot;}" data-execution_count="25">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">7</span>) </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>ticks <span class="op">=</span> [(c, s, <span class="ss">f'</span><span class="sc">{</span><span class="bu">round</span>(s<span class="op">/</span><span class="dv">1000000</span>, <span class="dv">2</span>)<span class="sc">}</span><span class="ss">'</span>) <span class="cf">for</span> c, l <span class="kw">in</span> utils.chrom_lengths[<span class="st">'hg38'</span>].items() <span class="cf">for</span> s <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, l, <span class="dv">2_500_000</span>)]</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>g.add_labels(ticks, base<span class="op">=</span>g.ideogram_base, min_height<span class="op">=</span>g.ideogram_height<span class="op">+</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">'black'</span>, zorder<span class="op">=</span><span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-26-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="chrom_ideograms_files/figure-html/cell-26-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Style label text and spine:</p>
<div id="cell-45" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:16.710713Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:16.710468Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:18.491123Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:18.490854Z&quot;}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord, color<span class="op">=</span><span class="st">'red'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-27-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="chrom_ideograms_files/figure-html/cell-27-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Remove <code>color</code> keyword arguemnt and instead add random label text and spine colors to the <code>sample_gene_coord</code> tuples:</p>
<div id="cell-47" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:18.493987Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:18.493863Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:18.518818Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:18.518556Z&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>rand_col <span class="op">=</span> <span class="kw">lambda</span> : matplotlib.colors.to_hex(plt.get_cmap(<span class="st">'hsv'</span>, <span class="dv">10</span>)(randint(<span class="dv">0</span>, <span class="dv">9</span>)))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>sample_gene_coord_with_colors <span class="op">=</span> [(<span class="op">*</span>t, rand_col()) <span class="cf">for</span> t <span class="kw">in</span> sample_gene_coord]</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>sample_gene_coord_with_colors[:<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>[('chr5', 181199559.5, 'TRIM7', '#ff00bf'),
 ('chr10', 1094378.0, 'WDR37', '#ff0000'),
 ('chr6', 12263854.5, 'EDN1', '#08ff00')]</code></pre>
</div>
</div>
<div id="cell-48" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:18.520423Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:18.520329Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:20.351448Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:20.351126Z&quot;}" data-execution_count="28">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord_with_colors,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>             base<span class="op">=</span>g.ideogram_base, min_height<span class="op">=</span>g.ideogram_height<span class="op">*</span><span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-29-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="chrom_ideograms_files/figure-html/cell-29-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Or with a single color scale to visually evaluate significance:</p>
<div id="cell-50" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:20.354252Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:20.354126Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:22.189081Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:22.188774Z&quot;}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>rand_col <span class="op">=</span> <span class="kw">lambda</span> : matplotlib.colors.to_hex(plt.get_cmap(<span class="st">'Blues'</span>, <span class="dv">10</span>)(randint(<span class="dv">0</span>, <span class="dv">9</span>)))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>sample_gene_coord_with_colors <span class="op">=</span> [(<span class="op">*</span>t, rand_col()) <span class="cf">for</span> t <span class="kw">in</span> sample_gene_coord]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.GenomeIdeogram(assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord_with_colors, linewidth<span class="op">=</span><span class="dv">1</span>, </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>             base<span class="op">=</span>g.ideogram_base, min_height<span class="op">=</span>g.ideogram_height<span class="op">*</span><span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-30-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="chrom_ideograms_files/figure-html/cell-30-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-51" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:22.191728Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:22.191600Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:22.357740Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:22.357378Z&quot;}" data-execution_count="30">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>rand_col <span class="op">=</span> <span class="kw">lambda</span> : matplotlib.colors.to_hex(plt.get_cmap(<span class="st">'viridis'</span>, <span class="dv">10</span>)(randint(<span class="dv">0</span>, <span class="dv">9</span>)))</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sample_gene_coord_with_colors <span class="op">=</span> [(<span class="op">*</span>t, rand_col()) <span class="cf">for</span> t <span class="kw">in</span> sample_gene_coord]</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chr2'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord_with_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-31-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="chrom_ideograms_files/figure-html/cell-31-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-52" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:22.359410Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:22.359309Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:22.531479Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:22.531142Z&quot;}" data-execution_count="31">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>random.seed(<span class="dv">4</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>rand_col <span class="op">=</span> <span class="kw">lambda</span> : matplotlib.colors.to_hex(plt.get_cmap(<span class="st">'Blues'</span>, <span class="dv">10</span>)(randint(<span class="dv">0</span>, <span class="dv">9</span>)))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>sample_gene_coord_with_sizes <span class="op">=</span> [(<span class="op">*</span>t, <span class="st">'black'</span>, random.random()) <span class="cf">for</span> t <span class="kw">in</span> sample_gene_coord]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chr2'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord_with_sizes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-32-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="chrom_ideograms_files/figure-html/cell-32-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="combinatorial-highlighting" class="level2">
<h2 class="anchored" data-anchor-id="combinatorial-highlighting">Combinatorial highlighting</h2>
<div id="cell-54" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:22.533247Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:22.533124Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:22.692809Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:22.692558Z&quot;}" data-execution_count="32">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chr2'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>             colored<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'CHST10'</span>, <span class="st">'CPS1'</span>],</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>             filled<span class="op">=</span>[<span class="st">'STAMBP'</span>, <span class="st">'SP140L'</span>],</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>             framed<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'C2orf88'</span>, <span class="st">'STAMBP'</span>],</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>             bold<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'SP140L'</span>, <span class="st">'EPC2'</span>],</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>             italic<span class="op">=</span>[<span class="st">'EPC2'</span>],</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>             )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-33-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="chrom_ideograms_files/figure-html/cell-33-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-55" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:22.694853Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:22.694717Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:22.931220Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:22.930939Z&quot;}" data-execution_count="33">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chr2'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>             colored<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'CHST10'</span>, <span class="st">'CPS1'</span>],</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>             filled<span class="op">=</span>[<span class="st">'STAMBP'</span>, <span class="st">'SP140L'</span>],</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>             framed<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'C2orf88'</span>, <span class="st">'STAMBP'</span>],</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>             bold<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'SP140L'</span>, <span class="st">'EPC2'</span>],</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>             italic<span class="op">=</span>[<span class="st">'EPC2'</span>],</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>             highlight_color<span class="op">=</span><span class="st">'darkorange'</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>             )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-34-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="chrom_ideograms_files/figure-html/cell-34-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>In combination with label sizes:</p>
<div id="cell-57" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:22.933029Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:22.932898Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:23.093573Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:23.093223Z&quot;}" data-execution_count="34">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chr2'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, font_size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord_with_sizes,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>             colored<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'CHST10'</span>, <span class="st">'CPS1'</span>],</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>             filled<span class="op">=</span>[<span class="st">'STAMBP'</span>, <span class="st">'SP140L'</span>],</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>             framed<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'C2orf88'</span>, <span class="st">'STAMBP'</span>],</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>             bold<span class="op">=</span>[<span class="st">'NCOA1'</span>, <span class="st">'SP140L'</span>, <span class="st">'EPC2'</span>],</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>             italic<span class="op">=</span>[<span class="st">'EPC2'</span>],</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>             )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-35-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="chrom_ideograms_files/figure-html/cell-35-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>You can do all the same things as with <code>GenomeIdeogram</code>:</p>
<div id="cell-59" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:23.095568Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:23.095435Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:23.267196Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:23.266918Z&quot;}" data-execution_count="35">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># g.draw_chromosomes(base=1, height=0.10, facecolor='lightgray')</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes()</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>g.add_labels(sample_gene_coord)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>g.add_segments([(<span class="st">'chrX'</span>, <span class="fl">25e6</span>, <span class="fl">55e6</span>), </span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                 (<span class="st">'chrX'</span>, <span class="fl">98e6</span>, <span class="fl">102e6</span>)], label<span class="op">=</span><span class="st">'segments'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>g.add_segments([(<span class="st">'chrX'</span>, <span class="fl">115e6</span>, <span class="fl">135e6</span>)], facecolor<span class="op">=</span><span class="st">'red'</span>, base<span class="op">=</span><span class="dv">4</span>, height<span class="op">=</span><span class="dv">1</span>, label<span class="op">=</span><span class="st">'red segments'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>g.add_segments([(<span class="st">'chrX'</span>, <span class="fl">125e6</span>, <span class="fl">145e6</span>)], facecolor<span class="op">=</span><span class="st">'green'</span>, base<span class="op">=</span><span class="dv">5</span>, height<span class="op">=</span><span class="dv">1</span>, label<span class="op">=</span><span class="st">'green segments'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>g.add_segments([(<span class="st">'chrX'</span>, <span class="fl">35e6</span>, <span class="fl">45e6</span>), </span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>                 (<span class="st">'chrX'</span>, <span class="fl">100e6</span>, <span class="fl">102e6</span>), </span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>                 (<span class="st">'chrX'</span>, <span class="fl">120e6</span>, <span class="fl">132e6</span>)], facecolor<span class="op">=</span><span class="st">'blue'</span>, base<span class="op">=</span><span class="dv">3</span>, height<span class="op">=</span><span class="dv">1</span>, label<span class="op">=</span><span class="st">'blue segments'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>g.add_vlines()</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="co"># plot using function that takes data and column name keyword args for dimensions (like seaborn's sns.scatterplot)</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>g.map_fun(sns.scatterplot, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, yaxis<span class="op">=</span>(<span class="fl">0.5</span>, <span class="fl">1.5</span>), s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'C1'</span>, label<span class="op">=</span><span class="st">'seaborn'</span>)</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a><span class="co"># plot using matplotlib Axes method that x and y positional arguments (like matplotlib's Axes.scatter)</span></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>g.map_method(Axes.scatter, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, yaxis<span class="op">=</span>(<span class="fl">1.5</span>, <span class="fl">2.5</span>), s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'C2'</span>, label<span class="op">=</span><span class="st">'matplotlib'</span>)</span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>g.add_legend()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-36-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="chrom_ideograms_files/figure-html/cell-36-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="additional-axes" class="level2">
<h2 class="anchored" data-anchor-id="additional-axes">Additional axes</h2>
<div id="cell-61" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:23.270425Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:23.270309Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:23.597066Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:23.596782Z&quot;}" data-execution_count="36">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>) <span class="co"># plots chromosome</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>, facecolor<span class="op">=</span><span class="st">'none'</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>new_ax, new_ax2, new_ax3 <span class="op">=</span> g.add_axes(<span class="dv">3</span>, height_ratio<span class="op">=</span><span class="fl">0.6</span>, hspace<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>_df <span class="op">=</span> utils.stairs(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>])</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>new_ax.plot(_df.pos, _df.y, label<span class="op">=</span><span class="st">'Stairs'</span>)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>new_ax.vlines(np.arange(<span class="op">*</span>new_ax.get_xlim(), <span class="dv">2_000_000</span>), <span class="op">*</span>new_ax.get_ylim(), color<span class="op">=</span><span class="st">'gray'</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>new_ax.hlines(np.arange(<span class="op">*</span>new_ax.get_ylim(), <span class="dv">1</span>), <span class="op">*</span>new_ax.get_xlim(), color<span class="op">=</span><span class="st">'gray'</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>new_ax.legend(loc<span class="op">=</span><span class="st">'center left'</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.02</span>, <span class="fl">0.5</span>), fontsize<span class="op">=</span><span class="st">"10"</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="op">;</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>new_ax2.scatter(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'x'</span>], df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'y'</span>], s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'black'</span>,label<span class="op">=</span><span class="st">'Scatter'</span>)</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>fit <span class="op">=</span> utils.fit_lowess(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>], <span class="st">'x'</span>, <span class="st">'y'</span>)</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>new_ax2.plot(fit.x, fit.y, c<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'Lowess'</span>)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>new_ax2.legend(loc<span class="op">=</span><span class="st">'center left'</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.02</span>, <span class="fl">0.5</span>), fontsize<span class="op">=</span><span class="st">"10"</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="op">;</span></span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>new_ax3.scatter(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'x'</span>], df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'y'</span>], s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'black'</span>,label<span class="op">=</span><span class="st">'Scatter'</span>)</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>fit <span class="op">=</span> utils.fit_lowess(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>], <span class="st">'x'</span>, <span class="st">'y'</span>, frac<span class="op">=</span><span class="fl">0.05</span>)</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>new_ax3.plot(fit.x, fit.y, c<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'Lowess'</span>)</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>new_ax3.legend(loc<span class="op">=</span><span class="st">'center left'</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.02</span>, <span class="fl">0.3</span>), fontsize<span class="op">=</span><span class="st">"10"</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-37-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="chrom_ideograms_files/figure-html/cell-37-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="floating-annotations" class="level2">
<h2 class="anchored" data-anchor-id="floating-annotations">Floating annotations</h2>
<div id="cell-63" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:23.600698Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:23.600580Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:23.857528Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:23.857199Z&quot;}" data-execution_count="37">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(<span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>) </span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="dv">2</span>) </span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>new_ax, new_ax2 <span class="op">=</span> g.add_axes(<span class="dv">2</span>, height_ratio<span class="op">=</span><span class="fl">0.6</span>, hspace<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>_df <span class="op">=</span> utils.stairs(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>])</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>new_ax.plot(_df.pos, _df.y, label<span class="op">=</span><span class="st">'Stairs'</span>)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>new_ax.legend(loc<span class="op">=</span><span class="st">'center left'</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.02</span>, <span class="fl">0.5</span>), </span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>              fontsize<span class="op">=</span><span class="st">"10"</span>, frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>new_ax2.scatter(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'x'</span>], df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>, <span class="st">'y'</span>], s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'black'</span>,label<span class="op">=</span><span class="st">'Scatter'</span>)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>fit <span class="op">=</span> utils.fit_lowess(df.loc[df.chrom <span class="op">==</span> <span class="st">'chrX'</span>], <span class="st">'x'</span>, <span class="st">'y'</span>)</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>new_ax2.plot(fit.x, fit.y, c<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'Lowess'</span>)</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>new_ax2.legend(loc<span class="op">=</span><span class="st">'center left'</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.02</span>, <span class="fl">0.5</span>), fontsize<span class="op">=</span><span class="st">"10"</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="op">;</span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>g.annotate((<span class="dv">0</span>, <span class="dv">4</span>), (<span class="dv">10000000</span>, <span class="dv">4</span>), </span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>         <span class="st">"This is the beginning of the chromosome"</span>, <span class="st">'chrX'</span>)</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>g.annotate((<span class="dv">61000000</span>, <span class="dv">4</span>), (<span class="dv">40000000</span>, <span class="dv">4</span>), </span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a>         <span class="st">"This is the centromere"</span>, <span class="st">'chrX'</span>)</span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>g.annotate((<span class="dv">90000000</span>, <span class="dv">1</span>), (<span class="dv">90000000</span>, <span class="dv">5</span>), </span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>         <span class="st">"Something important in another axes"</span>, <span class="st">'chrX'</span>, </span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>         ax<span class="op">=</span>new_ax2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-38-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="chrom_ideograms_files/figure-html/cell-38-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="zooms" class="level2">
<h2 class="anchored" data-anchor-id="zooms">Zooms</h2>
<div id="cell-65" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2025-02-09T10:26:23.860461Z&quot;,&quot;iopub.status.busy&quot;:&quot;2025-02-09T10:26:23.860349Z&quot;,&quot;iopub.status.idle&quot;:&quot;2025-02-09T10:26:24.346197Z&quot;,&quot;shell.execute_reply&quot;:&quot;2025-02-09T10:26:24.345806Z&quot;}" data-execution_count="38">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> gplt.ChromIdeogram(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'chrX'</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    zooms<span class="op">=</span>[</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>        (<span class="dv">45_000_000</span>, <span class="dv">55_000_000</span>),</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>        (<span class="dv">100_000_000</span>, <span class="dv">120_000_000</span>),</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>) </span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>g.draw_chromosomes(base<span class="op">=</span><span class="fl">3.5</span>, height<span class="op">=</span><span class="dv">1</span>, facecolor<span class="op">=</span><span class="st">'none'</span>)</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>g.add_horizon(data<span class="op">=</span>df, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>g.map_fun(sns.scatterplot, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, </span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>          yaxis<span class="op">=</span>(<span class="fl">0.5</span>, <span class="dv">1</span>), s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'C1'</span>, label<span class="op">=</span><span class="st">'seaborn'</span>)</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>g.map_method(Axes.scatter, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, </span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>             yaxis<span class="op">=</span>(<span class="fl">1.5</span>, <span class="fl">2.4</span>), s<span class="op">=</span><span class="dv">5</span>, c<span class="op">=</span><span class="st">'C2'</span>, label<span class="op">=</span><span class="st">'matplotlib'</span>)</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>g.map_method(Axes.plot, data<span class="op">=</span>utils.fit_lowess(df, <span class="st">'x'</span>, <span class="st">'y'</span>, frac<span class="op">=</span><span class="fl">0.02</span>), </span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>             x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span><span class="st">'y'</span>, yaxis<span class="op">=</span>(<span class="fl">1.5</span>, <span class="fl">2.4</span>), c<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'Lowess'</span>)</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>g.add_legend()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="chrom_ideograms_files/figure-html/cell-39-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="chrom_ideograms_files/figure-html/cell-39-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/munch-group\.github\.io\/geneinfo\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../pages/gene_plots.html" class="pagination-link" aria-label="Gene context">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Gene context</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/string_nets.html" class="pagination-link" aria-label="String networks">
        <span class="nav-page-text">String networks</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>