<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Example use â€“ geneinfo</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../api/" rel="next">
<link href="../pages/getting_started.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-659650fc26dc25888fc1474f317bb8ac.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../api/_styles-quartodoc.css">
<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">geneinfo</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../api/" aria-current="page"> 
<span class="menu-text">API reference</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/munch-group/geneinfo/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/examples.html">Examples</a></li><li class="breadcrumb-item"><a href="../pages/examples.html">Example use</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Examples</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/examples.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Example use</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../api/" class="sidebar-item-text sidebar-link">
 <span class="menu-text">API reference</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#summary-gene-information" id="toc-summary-gene-information" class="nav-link active" data-scroll-target="#summary-gene-information">Summary gene information</a></li>
  <li><a href="#get-the-hgcn-symbol-for-an-alias" id="toc-get-the-hgcn-symbol-for-an-alias" class="nav-link" data-scroll-target="#get-the-hgcn-symbol-for-an-alias">Get the HGCN symbol for an alias</a></li>
  <li><a href="#get-coordinates-for-genes-in-a-region" id="toc-get-coordinates-for-genes-in-a-region" class="nav-link" data-scroll-target="#get-coordinates-for-genes-in-a-region">Get coordinates for genes in a region</a></li>
  <li><a href="#plot-data-over-gene-annotation" id="toc-plot-data-over-gene-annotation" class="nav-link" data-scroll-target="#plot-data-over-gene-annotation">Plot data over gene annotation</a></li>
  <li><a href="#show-string-interaction-network" id="toc-show-string-interaction-network" class="nav-link" data-scroll-target="#show-string-interaction-network">Show STRING interaction network</a></li>
  <li><a href="#chromosome-ideogram" id="toc-chromosome-ideogram" class="nav-link" data-scroll-target="#chromosome-ideogram">Chromosome ideogram</a></li>
  <li><a href="#gene-ontology-information" id="toc-gene-ontology-information" class="nav-link" data-scroll-target="#gene-ontology-information">Gene ontology information</a></li>
  <li><a href="#go-graphs" id="toc-go-graphs" class="nav-link" data-scroll-target="#go-graphs">GO graphs</a></li>
  <li><a href="#find-genes-based-on-go-info" id="toc-find-genes-based-on-go-info" class="nav-link" data-scroll-target="#find-genes-based-on-go-info">Find genes based on GO info</a></li>
  <li><a href="#global-go-enrichment" id="toc-global-go-enrichment" class="nav-link" data-scroll-target="#global-go-enrichment">Global GO enrichment</a></li>
  <li><a href="#enrichment-among-a-subset-of-go-terms" id="toc-enrichment-among-a-subset-of-go-terms" class="nav-link" data-scroll-target="#enrichment-among-a-subset-of-go-terms">Enrichment among a subset of GO terms</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/examples.html">Examples</a></li><li class="breadcrumb-item"><a href="../pages/examples.html">Example use</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Example use</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="cell-2" class="cell" data-tags="[]">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys, os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib_inline.backend_inline <span class="im">import</span> set_matplotlib_formats</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>set_matplotlib_formats(<span class="st">'retina'</span>, <span class="st">'png'</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.patches <span class="im">import</span> Rectangle, Polygon</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>sns.set_style(<span class="st">'white'</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo <span class="im">as</span> gi</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>gi.email(<span class="st">'your@email.com'</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co">#%run -i build_cache.py</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>In-house uliity for accessing gene lists in google spreadsheet:</strong></p>
<div id="cell-4" class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> math <span class="im">import</span> sqrt</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> itertools <span class="im">import</span> zip_longest</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> list_genes(words, ncols<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    n <span class="op">=</span> <span class="bu">len</span>(words)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    col_width <span class="op">=</span> <span class="bu">max</span>(<span class="bu">map</span>(<span class="bu">len</span>, words)) <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> ncols <span class="kw">is</span> <span class="va">None</span>:</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>        ncols <span class="op">=</span> <span class="bu">max</span>(<span class="dv">100</span><span class="op">//</span>col_width, <span class="dv">1</span><span class="op">+</span>sqrt(n<span class="op">/</span>col_width))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    nrows <span class="op">=</span> <span class="bu">int</span>(n<span class="op">/</span>ncols) <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    rows <span class="op">=</span> []</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, n, nrows):</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>        rows.append(words[r:r<span class="op">+</span>nrows])</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> row <span class="kw">in</span> <span class="bu">list</span>(zip_longest(<span class="op">*</span>rows, fillvalue<span class="op">=</span><span class="st">''</span>)):</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>        line <span class="op">=</span> []</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> gene <span class="kw">in</span> row:</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>            line.append(gene.ljust(col_width))</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">''</span>.join(line))</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> read_google_sheet():</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    SHEET_ID <span class="op">=</span> <span class="st">'1JSjSLuto3jqdEnnG7JqzeC_1pUZw76n7XueVAYrUOpk'</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    SHEET_NAME <span class="op">=</span> <span class="st">'Sheet1'</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>    url <span class="op">=</span> <span class="ss">f'https://docs.google.com/spreadsheets/d/</span><span class="sc">{</span>SHEET_ID<span class="sc">}</span><span class="ss">/gviz/tq?tqx=out:csv&amp;sheet=</span><span class="sc">{</span>SHEET_NAME<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> pd.read_csv(url, header<span class="op">=</span><span class="dv">1</span>, low_memory<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df.loc[:, [<span class="kw">not</span> x.startswith(<span class="st">'Unnamed'</span>) <span class="cf">for</span> x <span class="kw">in</span> df.columns]]</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> gene_list_names():</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> read_google_sheet()</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">sorted</span>(df.columns.tolist())</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> gene_list(name):</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> read_google_sheet()</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>    sr <span class="op">=</span> df[name]</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> sr[<span class="op">~</span>sr.isnull()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-5" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># from pkg_resources import resource_filename</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># with open(resource_filename('geneinfo', 'data/gene_lists.pickle'), 'rb') as f:</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#     gene_lists = pickle.load(f)</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># def gene_list_names():</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">#     return sorted(gene_lists.keys())</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># def gene_list(name):</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co">#     return gene_lists[name].tolist()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-6" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>gene_list_names()        </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>['accel_reg_simiiformes_br',
 'all_npx',
 'all_post_mei_expr',
 'ari_all',
 'ari_nonPUR',
 'ari_relate_AFR',
 'ari_relate_ASIA',
 'ari_relate_EUR',
 'ari_relate_PUR',
 'cDEG',
 'candidates',
 'ech75_regions',
 'ech90_regions',
 'expr_mod_xi_copynr_fibrobl',
 'expr_mod_xi_copynr_lcl',
 'gametologs',
 'hpa_brain_prot',
 'intel_seiz_lang',
 'intelect_disabil',
 'intersection(nDEG, matos_neuron)',
 'matos_common',
 'matos_common_no_xy_gametologs',
 'matos_neuron',
 'meritxell_spermatid_expr',
 'msci',
 'msci_esc',
 'mult_copy',
 'my_primate_codeml',
 'nDEG',
 'neuron_genome_proteome',
 'neuron_npx_proteome',
 'primate_ampl_multi',
 'reg_sa_pheno',
 'sfari_all_conf',
 'union(H,matos_common)',
 'xi',
 'xi_any_evidence',
 'xi_escape',
 'xi_uncertain']</code></pre>
</div>
</div>
<div id="cell-7" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>xi_escape <span class="op">=</span> gene_list(<span class="st">'xi_escape'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>list_genes(xi_escape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AFF2    CDK16   EFNB1   GPR173  IKBKG   MAMLD1  OFD1    PPP2R3B SEPTIN6 TCEAL9  XG      
AKAP17A CDR1    EIF1AX  GPR174  IL3RA   MAP7D2  P2RY10  PQBP1   SLC25A6 TCEANC  XIST    
ALG13   CHM     EIF2S3  GTPBP6  IQSEC2  MED14   P2RY8   PRKX    SMC1A   TRAPPC2 XPNPEP2 
AP1S2   COX7B   FGF13   GYG2    JPX     MSL3    PDZD11  PUDP    SRPX    TXLNG   ZBED1   
ARSD    CXorf38 FHL1    HCFC1   KDM5C   MXRA5   PHKA1   RAB9A   STK26   UBA1    ZFX     
ASMTL   CXorf65 FOXO4   HEPH    KDM6A   NAA10   PLCXD1  RBBP7   STS     UBQLN2  ZRSR2   
CA5B    DDX3X   FUNDC1  HMGN5   LRCH2   NHS     PNPLA4  RPS4X   SYAP1   USP9X           
CD99    DHRSX   GEMIN8  HUWE1   MAGEE1  NHSL2   PORCN   SCML2   TASL    WAS             </code></pre>
</div>
</div>
<p><strong>Left align in pandas tables:</strong></p>
<div id="cell-9" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> left:</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__rlshift__</span>(<span class="va">self</span>, df):</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"Left align columns of data frame: df &lt;&lt; left()"</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>        left_aligned_df <span class="op">=</span> df.style.set_properties(<span class="op">**</span>{<span class="st">'text-align'</span>: <span class="st">'left'</span>})</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>        left_aligned_df <span class="op">=</span> left_aligned_df.set_table_styles(</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>        [<span class="bu">dict</span>(selector <span class="op">=</span> <span class="st">'th'</span>, props<span class="op">=</span>[(<span class="st">'text-align'</span>, <span class="st">'left'</span>)])])</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>        display(left_aligned_df)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(<span class="bu">dict</span>(name<span class="op">=</span>[<span class="st">'A'</span>, <span class="st">'B'</span>], desc<span class="op">=</span>[<span class="st">'bla bla bla bla'</span>, <span class="st">'bla'</span>]))</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">desc</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>A</td>
<td>bla bla bla bla</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>B</td>
<td>bla</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-10" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">&lt;&lt;</span> left()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style type="text/css">
#T_58978 th {
  text-align: left;
}
#T_58978_row0_col0, #T_58978_row0_col1, #T_58978_row1_col0, #T_58978_row1_col1 {
  text-align: left;
}
</style>

<table id="T_58978" class="caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_58978_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">name</th>
<th id="T_58978_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">desc</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_58978_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</td>
<td id="T_58978_row0_col0" class="data row0 col0">A</td>
<td id="T_58978_row0_col1" class="data row0 col1">bla bla bla bla</td>
</tr>
<tr class="even">
<td id="T_58978_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</td>
<td id="T_58978_row1_col0" class="data row1 col0">B</td>
<td id="T_58978_row1_col1" class="data row1 col1">bla</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-11" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>html</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>style<span class="op">&gt;</span> table { font<span class="op">-</span>variant<span class="op">-</span>numeric: tabular<span class="op">-</span>nums<span class="op">;</span> } <span class="op">&lt;/</span>style<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style> table { font-variant-numeric: tabular-nums; } </style>
</div>
</div>
<section id="summary-gene-information" class="level2">
<h2 class="anchored" data-anchor-id="summary-gene-information">Summary gene information</h2>
<p><strong>Using gene symbol:</strong></p>
<div id="cell-14" class="cell" data-tags="[]" data-execution_count="17">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>gi.gene_info(<span class="st">'USP4'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>USP4</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> UNP, Unph<br>
<em>ubiquitin specific peptidase 4</em><br>
<strong>Summary:</strong> The protein encoded by this gene is a protease that deubiquitinates target proteins such as ADORA2A and TRIM21. The encoded protein shuttles between the nucleus and cytoplasm and is involved in maintaining operational fidelity in the endoplasmic reticulum. Three transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Oct 2011].<br>
<strong>Genomic position:</strong> 3:49277144-49340712 (hg38), 3:49315264-49378145 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=USP4">Gene card</a></p>
<hr>
</div>
</div>
<p><strong>Using NCBI ID:</strong></p>
<div id="cell-16" class="cell" data-tags="[]" data-execution_count="18">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>gi.gene_info(<span class="dv">7157</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>TP53</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> BCC7, BMFS5, LFS1, P53, TRP53<br>
<em>tumor protein p53</em><br>
<strong>Summary:</strong> This gene encodes a tumor suppressor protein containing transcriptional activation, DNA binding, and oligomerization domains. The encoded protein responds to diverse cellular stresses to regulate expression of target genes, thereby inducing cell cycle arrest, apoptosis, senescence, DNA repair, or changes in metabolism. Mutations in this gene are associated with a variety of human cancers, including hereditary cancers such as Li-Fraumeni syndrome. Alternative splicing of this gene and the use of alternate promoters result in multiple transcript variants and isoforms. Additional isoforms have also been shown to result from the use of alternate translation initiation codons from identical transcript variants (PMIDs: 12032546, 20937277). [provided by RefSeq, Dec 2016].<br>
<strong>Genomic position:</strong> 17:7661779-7687546 (hg38), 17:7565097-7590856 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=TP53">Gene card</a></p>
<hr>
</div>
</div>
<p><strong>For a a list of genes:</strong></p>
<div id="cell-18" class="cell" data-tags="[]" data-execution_count="19">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>gi.gene_info([<span class="st">'LARS2'</span>, <span class="st">'XCR1'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>LARS2</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> HLASA, LEURS, PRLTS4, mtLeuRS<br>
<em>leucyl-tRNA synthetase 2, mitochondrial</em><br>
<strong>Summary:</strong> This gene encodes a class 1 aminoacyl-tRNA synthetase, mitochondrial leucyl-tRNA synthetase. Each of the twenty aminoacyl-tRNA synthetases catalyzes the aminoacylation of a specific tRNA or tRNA isoaccepting family with the cognate amino acid. [provided by RefSeq, Jul 2008].<br>
<strong>Genomic position:</strong> 3:45388561-45554726 (hg38), 3:45429998-45590913 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=LARS2">Gene card</a></p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>XCR1</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> CCXCR1, GPR5<br>
<em>X-C motif chemokine receptor 1</em><br>
<strong>Summary:</strong> The protein encoded by this gene is a chemokine receptor belonging to the G protein-coupled receptor superfamily. The family members are characterized by the presence of 7 transmembrane domains. The encoded protein transduces a signal by increasing the intracellular calcium ion level. The viral macrophage inflammatory protein-II is an antagonist of this receptor and blocks signaling. Some studies have implicated a cluster of genes at 3p21.31, including this gene, as associated with COVID-19 risk. The encoded protein may also play a role in cell proliferation and migration in several types of cancer. [provided by RefSeq, Jan 2023].<br>
<strong>Genomic position:</strong> 3:46016990-46085825 (hg38), 3:46058516-46069234 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=XCR1">Gene card</a></p>
<hr>
</div>
</div>
<p><strong>For genes a region:</strong></p>
<div id="cell-20" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gi.gene_info_region(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77000000</span>, assembly<span class="op">=</span><span class="st">'hg19'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>FGF16</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> FGF-16, MF4<br>
<em>fibroblast growth factor 16</em><br>
<strong>Summary:</strong> This gene encodes a member of a family of proteins that are involved in a variety of biological processes, including embryonic development, cell growth, morphogenesis, tissue repair, tumor growth and invasion. This gene is expressed in cardiac cells and is required for proper heart development. Mutation in this gene was also observed in individuals with metacarpal 4-5 fusion. [provided by RefSeq, Mar 2014].<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=FGF16">Gene card</a></p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>ATRX</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> JMS, MRX52, RAD54, RAD54L, XH2, XNP, ZNF-HX<br>
<em>ATRX chromatin remodeler</em><br>
<strong>Summary:</strong> The protein encoded by this gene contains an ATPase/helicase domain, and thus it belongs to the SWI/SNF family of chromatin remodeling proteins. This protein is found to undergo cell cycle-dependent phosphorylation, which regulates its nuclear matrix and chromatin association, and suggests its involvement in the gene regulation at interphase and chromosomal segregation in mitosis. Mutations in this gene are associated with X-linked syndromes exhibiting cognitive disabilities as well as alpha-thalassemia (ATRX) syndrome. These mutations have been shown to cause diverse changes in the pattern of DNA methylation, which may provide a link between chromatin remodeling, DNA methylation, and gene expression in developmental processes. Multiple alternatively spliced transcript variants encoding distinct isoforms have been reported. [provided by RefSeq, Jul 2017].<br>
<strong>Genomic position:</strong> X:77504880-77786233 (hg38), HG1426_PATCH:76765832-77047170, X:76760356-77041702 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=ATRX">Gene card</a></p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>ATRX</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> JMS, MRX52, RAD54, RAD54L, XH2, XNP, ZNF-HX<br>
<em>ATRX chromatin remodeler</em><br>
<strong>Summary:</strong> The protein encoded by this gene contains an ATPase/helicase domain, and thus it belongs to the SWI/SNF family of chromatin remodeling proteins. This protein is found to undergo cell cycle-dependent phosphorylation, which regulates its nuclear matrix and chromatin association, and suggests its involvement in the gene regulation at interphase and chromosomal segregation in mitosis. Mutations in this gene are associated with X-linked syndromes exhibiting cognitive disabilities as well as alpha-thalassemia (ATRX) syndrome. These mutations have been shown to cause diverse changes in the pattern of DNA methylation, which may provide a link between chromatin remodeling, DNA methylation, and gene expression in developmental processes. Multiple alternatively spliced transcript variants encoding distinct isoforms have been reported. [provided by RefSeq, Jul 2017].<br>
<strong>Genomic position:</strong> X:77504880-77786233 (hg38), HG1426_PATCH:76765832-77047170, X:76760356-77041702 (hg19)<br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=ATRX">Gene card</a></p>
<hr>
</div>
</div>
<p>For species other than human:</p>
<div id="cell-22" class="cell" data-tags="[]" data-execution_count="21">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>gi.gene_info(<span class="st">'DYNLT3'</span>, species<span class="op">=</span><span class="st">'mouse'</span>, scopes<span class="op">=</span>[<span class="st">'mgi'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Symbol:</strong> <strong><em>Dynlt3</em></strong> (protein-coding) &nbsp; &nbsp; &nbsp; &nbsp; <strong>Aliases:</strong> 2310075M16Rik, Tcte1l<br>
<em>dynein light chain Tctex-type 3</em><br>
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=Dynlt3">Gene card</a></p>
<hr>
</div>
</div>
</section>
<section id="get-the-hgcn-symbol-for-an-alias" class="level2">
<h2 class="anchored" data-anchor-id="get-the-hgcn-symbol-for-an-alias">Get the HGCN symbol for an alias</h2>
<div id="cell-24" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>gi.hgcn_symbol([<span class="st">'CXorf23'</span>, <span class="st">'BCLAF3'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>['BCLAF3', 'BCLAF3']</code></pre>
</div>
</div>
</section>
<section id="get-coordinates-for-genes-in-a-region" class="level2">
<h2 class="anchored" data-anchor-id="get-coordinates-for-genes-in-a-region">Get coordinates for genes in a region</h2>
<div id="cell-26" class="cell" data-tags="[]" data-execution_count="4">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gi.get_genes_region(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77000000</span>, assembly<span class="op">=</span><span class="st">'hg19'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>[('FGF16',
  76709642,
  76712769,
  '+',
  [(76709642, 76709751), (76711767, 76712769)]),
 ('ATRX',
  76760357,
  77041702,
  '-',
  [(76760357, 76764107),
   (76776265, 76776394),
   (76776880, 76776976),
   (76777740, 76777866),
   (76778729, 76778879),
   (76812921, 76813116),
   (76814139, 76814317),
   (76829714, 76829823),
   (76845303, 76845410),
   (76849165, 76849319),
   (76854879, 76855049),
   (76855200, 76855289),
   (76855902, 76856033),
   (76872080, 76872198),
   (76874273, 76874449),
   (76875862, 76876000),
   (76888694, 76888872),
   (76889053, 76889200),
   (76890084, 76890194),
   (76891405, 76891547),
   (76907603, 76907843),
   (76909587, 76909690),
   (76912049, 76912143),
   (76918870, 76919047),
   (76920133, 76920267),
   (76931720, 76931793),
   (76937011, 76940085),
   (76940430, 76940498),
   (76944310, 76944420),
   (76949312, 76949426),
   (76952064, 76952192),
   (76953070, 76953123),
   (76954061, 76954117),
   (76972607, 76972720),
   (77041467, 77041702)]),
 ('ATRX',
  76760357,
  77041702,
  '-',
  [(76760357, 76764107),
   (76776265, 76776394),
   (76776880, 76776976),
   (76777740, 76777866),
   (76778729, 76778879),
   (76812921, 76813116),
   (76814139, 76814317),
   (76829714, 76829823),
   (76845303, 76845410),
   (76849165, 76849319),
   (76854879, 76855049),
   (76855200, 76855289),
   (76855902, 76856033),
   (76872080, 76872198),
   (76874273, 76874449),
   (76875862, 76876000),
   (76888694, 76888872),
   (76889053, 76889200),
   (76890084, 76890194),
   (76891405, 76891547),
   (76907603, 76907843),
   (76909587, 76909690),
   (76912049, 76912143),
   (76918870, 76919047),
   (76920133, 76920267),
   (76931720, 76931793),
   (76937011, 76940085),
   (76940430, 76940498),
   (76944310, 76944420),
   (76952064, 76952192),
   (76953070, 76953123),
   (76954061, 76954117),
   (76972607, 76972720),
   (77041467, 77041702)])]</code></pre>
</div>
</div>
</section>
<section id="plot-data-over-gene-annotation" class="level2">
<h2 class="anchored" data-anchor-id="plot-data-over-gene-annotation">Plot data over gene annotation</h2>
<div id="cell-28" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># x = np.linspace(1, 1520300000, 100000)</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># y = np.random.random(100000)</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dummy manhattan data</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> np.linspace(<span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="dv">10000</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>rng <span class="op">=</span> np.random.default_rng()</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> rng.exponential(scale<span class="op">=</span><span class="dv">1</span>, size<span class="op">=</span>x.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Standard plot:</strong></p>
<div id="cell-30" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="st">'hg38'</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-18-output-1.png" width="630" height="472" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Same on older hg19 assembly:</strong></p>
<div id="cell-32" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, assembly<span class="op">=</span><span class="st">'hg19'</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-19-output-1.png" width="629" height="472" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Two subplots for data:</strong></p>
<div id="cell-34" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>ax1, ax2 <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, return_axes<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>ax1.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>ax2.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-20-output-1.png" width="630" height="472" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Three subplots for data and adjusted aspect ratio:</strong></p>
<div id="cell-36" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ax1, ax2, ax3 <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, assembly<span class="op">=</span><span class="st">'hg38'</span>, return_axes<span class="op">=</span><span class="dv">3</span>, aspect<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>ax1.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>ax2.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>ax3.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-21-output-1.png" width="630" height="472" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Highlight a list of genes:</strong></p>
<div id="cell-38" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="st">'hg19'</span>, highlight<span class="op">=</span>[<span class="st">'ATRX'</span>, <span class="st">'ATP7A'</span>])</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax) <span class="co"># make it look nicer</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-22-output-1.png" width="629" height="471" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Combinatorial hightlighting:</strong></p>
<p>Combining text weight, style, color and bbox:</p>
<div id="cell-40" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>highlight <span class="op">=</span> defaultdict(<span class="bu">dict</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> [<span class="st">'ATRX'</span>, <span class="st">'COX7B'</span>, <span class="st">'PGK1'</span>]:</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(weight<span class="op">=</span><span class="st">'bold'</span>))</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> [<span class="st">'MAGT1'</span>]:</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(style<span class="op">=</span><span class="st">'italic'</span>))</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> [<span class="st">'ATP7A'</span>, <span class="st">'ATRX'</span>]:</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(color<span class="op">=</span><span class="st">'red'</span>))</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> [<span class="st">'PGK1'</span>, <span class="st">'ATRX'</span>]:</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(bbox<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">'black'</span>, facecolor<span class="op">=</span><span class="st">'none'</span>, linewidth<span class="op">=</span><span class="fl">0.2</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-41" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="st">'hg19'</span>, highlight<span class="op">=</span>highlight)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax) <span class="co"># make it look nicer</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-24-output-1.png" width="629" height="471" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-42" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>highlight <span class="op">=</span> defaultdict(<span class="bu">dict</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> gene_list(<span class="st">'xi_any_evidence'</span>):</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(weight<span class="op">=</span><span class="st">'bold'</span>))</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> gene_list(<span class="st">'ech75_regions'</span>):</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(style<span class="op">=</span><span class="st">'italic'</span>))</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> gene_list(<span class="st">'sfari_all_conf'</span>) <span class="op">+</span> gene_list(<span class="st">'intelect_disabil'</span>):</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(color<span class="op">=</span><span class="st">'red'</span>))</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> gene_list(<span class="st">'gametologs'</span>) <span class="op">+</span>  gene_list(<span class="st">'mult_copy'</span>) <span class="op">+</span> gene_list(<span class="st">'xi_escape'</span>) <span class="op">+</span> gene_list(<span class="st">'cDEG'</span>) <span class="op">+</span> gene_list(<span class="st">'nDEG'</span>):</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    highlight[gene].update(<span class="bu">dict</span>(bbox<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">'black'</span>, facecolor<span class="op">=</span><span class="st">'none'</span>, linewidth<span class="op">=</span><span class="fl">0.2</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-43" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="st">'hg19'</span>, highlight<span class="op">=</span>highlight)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax) <span class="co"># make it look nicer</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-26-output-1.png" width="629" height="471" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Larger with colors:</strong></p>
<div id="cell-45" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, <span class="st">'hg19'</span>, highlight<span class="op">=</span>highlight, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>, c<span class="op">=</span>y, cmap<span class="op">=</span><span class="st">'viridis_r'</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax) <span class="co"># make it look nicer</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-27-output-1.png" width="989" height="392" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Other features:</strong></p>
<div id="cell-47" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gi.gene_plot(<span class="st">'chrX'</span>, <span class="dv">76500000</span>, <span class="dv">77500000</span>, assembly<span class="op">=</span><span class="st">'hg19'</span>,</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>                  figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>),          <span class="co"># larger figure</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                  collapse_splice_var<span class="op">=</span><span class="va">False</span>, <span class="co"># show all splice variants separately</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>                  exact_exons<span class="op">=</span><span class="va">False</span>,        <span class="co"># allow exons to be wide enough to be visible on large plots</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>                  hard_limits<span class="op">=</span><span class="va">True</span>,          <span class="co"># truncate genes that overlap the borders of the specified range</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>                 )</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, s<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax) <span class="co"># make it look nicer</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-28-output-1.png" width="989" height="392" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="show-string-interaction-network" class="level2">
<h2 class="anchored" data-anchor-id="show-string-interaction-network">Show STRING interaction network</h2>
<div id="cell-49" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>gi.show_string_network([<span class="st">'ATRX'</span>, <span class="st">'TP53'</span>, <span class="st">'RAD51'</span>], nodes<span class="op">=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="52">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-29-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="chromosome-ideogram" class="level2">
<h2 class="anchored" data-anchor-id="chromosome-ideogram">Chromosome ideogram</h2>
<div id="cell-51" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import importlib</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># import geneinfo</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># importlib.reload(geneinfo)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-52" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>annotation <span class="op">=</span> [</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    (<span class="st">'chr1'</span>, <span class="dv">50000000</span>, <span class="dv">50001000</span>),                          <span class="co"># segment color defaults to red</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    (<span class="st">'chr2'</span>, <span class="dv">10000000</span>, <span class="dv">1000000</span>, <span class="st">'green'</span>),                  <span class="co"># with segment color</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    (<span class="st">'chr3'</span>, <span class="dv">45000000</span>, <span class="dv">55000000</span>, <span class="st">'blue'</span>, <span class="st">'NAME'</span>),          <span class="co"># with name label</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    (<span class="st">'chr6'</span>, <span class="dv">10000000</span>, <span class="dv">15000000</span>, <span class="st">'magenta'</span>, <span class="va">None</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="co"># with base and height</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    (<span class="st">'chr6'</span>, <span class="dv">10000000</span>, <span class="dv">15000000</span>, <span class="st">'pink'</span>, <span class="va">None</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>),  <span class="co"># with base and height</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>gi.chrom_ideogram(annotation)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-31-output-1.png" width="813" height="854" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>To make sure segments are visible on the plot, the default minimal size of plotteed segments are 200000. This can be changed like this (notice how the chr1 segment is no longer visible):</p>
<div id="cell-54" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>gi.chrom_ideogram(annotation, min_visible_width<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-32-output-1.png" width="813" height="854" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Other assembly, figure dimentions:</p>
<div id="cell-56" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>gi.chrom_ideogram(annotation, assembly<span class="op">=</span><span class="st">'hg19'</span>, figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-33-output-1.png" width="345" height="700" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>You can add position markers along a chromosome like this to better see the location of some segments:</strong></p>
<div id="cell-58" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>chrom <span class="op">=</span> <span class="st">'chr3'</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>start <span class="op">=</span> <span class="dv">100000000</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>end <span class="op">=</span> <span class="dv">200000000</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>width <span class="op">=</span> <span class="dv">10000000</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>annotation2 <span class="op">=</span> annotation <span class="op">+</span> [(chrom, x[<span class="dv">0</span>], x[<span class="dv">1</span>], <span class="st">'black'</span>, <span class="bu">str</span>(x[<span class="dv">2</span>]<span class="op">/</span>width)) <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">zip</span>(</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">range</span>(start, end, width), </span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">range</span>(start<span class="op">+</span><span class="dv">1</span>, end, width), </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">range</span>(start, end, width)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>)]</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>gi.chrom_ideogram(annotation2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-34-output-1.png" width="813" height="854" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="gene-ontology-information" class="level2">
<h2 class="anchored" data-anchor-id="gene-ontology-information">Gene ontology information</h2>
<p><strong>NB:</strong> The first time you use the GO utilities, it will spend a <em>very</em> long time downloading data. If you do not want to wait that out in the notebook, you can execute this command in the folder where you store your notebooks:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode txt code-with-copy"><code class="sourceCode default"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>conda run -n &lt;environment&gt; python -c 'import geneinfo ; geneinfo.download_data()'</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Get info for one or more GO terms:</strong></p>
<div id="cell-62" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>gi.go_info(<span class="st">'GO:0000776'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0000776:</span></strong> <strong>kinetochore</strong><br>
A multisubunit complex that is located at the centromeric region of DNA and provides an attachment point for the spindle microtubules.</p>
<hr>
</div>
</div>
<div id="cell-63" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>gi.go_info([<span class="st">'GO:0000776'</span>, <span class="st">'GO:0005634'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0000776:</span></strong> <strong>kinetochore</strong><br>
A multisubunit complex that is located at the centromeric region of DNA and provides an attachment point for the spindle microtubules.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005634:</span></strong> <strong>nucleus</strong><br>
A membrane-bounded organelle of eukaryotic cells in which chromosomes are housed and replicated. In most cells, the nucleus contains all of the cellâ€™s chromosomes except the organellar chromosomes, and is the site of RNA synthesis and processing. In some species, or in specialized cell types, RNA metabolism or DNA replication may be absent.</p>
<hr>
</div>
</div>
<p><strong>Get GO terms for a one or more genes:</strong></p>
<div id="cell-65" class="cell" data-tags="[]" data-execution_count="43">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>go_terms <span class="op">=</span> gi.get_go_terms_for_genes([<span class="st">'DYNLT3'</span>])</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>go_terms[:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="43">
<pre><code>['GO:0000776',
 'GO:0005515',
 'GO:0005634',
 'GO:0005737',
 'GO:0005868',
 'GO:0007018',
 'GO:0007346',
 'GO:0042802',
 'GO:0045505',
 'GO:0045931']</code></pre>
</div>
</div>
<p><strong>Descriptions for GO terms for one or more genes:</strong></p>
<div id="cell-67" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>gi.go_info(gi.get_go_terms_for_genes([<span class="st">'DYNLT3'</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0000776:</span></strong> <strong>kinetochore</strong><br>
A multisubunit complex that is located at the centromeric region of DNA and provides an attachment point for the spindle microtubules.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005515:</span></strong> <strong>protein binding</strong><br>
Binding to a protein.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005634:</span></strong> <strong>nucleus</strong><br>
A membrane-bounded organelle of eukaryotic cells in which chromosomes are housed and replicated. In most cells, the nucleus contains all of the cellâ€™s chromosomes except the organellar chromosomes, and is the site of RNA synthesis and processing. In some species, or in specialized cell types, RNA metabolism or DNA replication may be absent.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005737:</span></strong> <strong>cytoplasm</strong><br>
The contents of a cell excluding the plasma membrane and nucleus, but including other subcellular structures.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005868:</span></strong> <strong>cytoplasmic dynein complex</strong><br>
Any dynein complex with a homodimeric dynein heavy chain core that catalyzes movement along a microtubule. Cytoplasmic dynein complexes participate in many cytoplasmic transport activities in eukaryotes, such as mRNA localization, intermediate filament transport, nuclear envelope breakdown, apoptosis, transport of centrosomal proteins, mitotic spindle assembly, virus transport, kinetochore functions, and movement of signaling and spindle checkpoint proteins. Some complexes participate in intraflagellar transport. Subunits associated with the dynein heavy chain mediate association between dynein heavy chain and cargoes, and may include light chains and light intermediate chains.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0007018:</span></strong> <strong>microtubule-based movement</strong><br>
A microtubule-based process that results in the movement of organelles, other microtubules, or other cellular components. Examples include motor-driven movement along microtubules and movement driven by polymerization or depolymerization of microtubules.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0007346:</span></strong> <strong>regulation of mitotic cell cycle</strong><br>
Any process that modulates the rate or extent of progress through the mitotic cell cycle.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0042802:</span></strong> <strong>identical protein binding</strong><br>
Binding to an identical protein or proteins.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0045505:</span></strong> <strong>dynein intermediate chain binding</strong><br>
Binding to an intermediate chain of the dynein complex.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0045931:</span></strong> <strong>positive regulation of mitotic cell cycle</strong><br>
Any process that activates or increases the rate or extent of progression through the mitotic cell cycle.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0051301:</span></strong> <strong>cell division</strong><br>
The process resulting in division and partitioning of components of a cell to form more cells; may or may not be accompanied by the physical separation of a cell into distinct, individually membrane-bounded daughter cells.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0061673:</span></strong> <strong>mitotic spindle astral microtubule</strong><br>
Any of the mitotic spindle microtubules that radiate in all directions from the spindle poles and are thought to contribute to the forces that separate the poles and position them in relation to the rest of the cell.</p>
<hr>
</div>
</div>
</section>
<section id="go-graphs" class="level2">
<h2 class="anchored" data-anchor-id="go-graphs">GO graphs</h2>
<p><strong>Show the sub-graph for a list of GO terms:</strong></p>
<div id="cell-70" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> importlib</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>importlib.<span class="bu">reload</span>(geneinfo)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co">#geneinfo.go_info('GO:0000776')</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>geneinfo.show_go_dag_for_terms(go_terms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="48">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-39-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-71" class="cell" data-tags="[]" data-execution_count="49">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_terms(go_terms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="49">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-40-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Show the sub-graph for the GO terms of a single gene:</strong></p>
<div id="cell-73" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_gene(<span class="st">'DYNLT3'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-41-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Same but using GO annotation backed by certain kinds of evidence:</strong></p>
<div id="cell-75" class="cell" data-tags="[]" data-execution_count="51">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_gene(<span class="st">'DYNLT3'</span>, evidence<span class="op">=</span>[<span class="st">'IDA'</span>, <span class="st">'EXP'</span>, <span class="st">'HTP'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="51">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-42-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Show the categories of evidence backing GO annotations:</strong></p>
<div id="cell-77" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_evidence_codes()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong>Experimental evidence codes:</strong> <br> Inferred from Experiment (EXP) <br> Inferred from Direct Assay (IDA) <br> Inferred from Physical Interaction (IPI) <br> Inferred from Mutant Phenotype (IMP) <br> Inferred from Genetic Interaction (IGI) <br> Inferred from Expression Pattern (IEP) <br> Inferred from High Throughput Experiment (HTP) <br> Inferred from High Throughput Direct Assay (HDA) <br> Inferred from High Throughput Mutant Phenotype (HMP) <br> Inferred from High Throughput Genetic Interaction (HGI) <br> Inferred from High Throughput Expression Pattern (HEP)</p>
<p><strong>Phylogenetically-inferred annotations:</strong> <br> Inferred from Biological aspect of Ancestor (IBA) <br> Inferred from Biological aspect of Descendant (IBD) <br> Inferred from Key Residues (IKR) <br> Inferred from Rapid Divergence (IRD)</p>
<p><strong>Computational analysis evidence codes</strong> <br> Inferred from Sequence or structural Similarity (ISS) <br> Inferred from Sequence Orthology (ISO) <br> Inferred from Sequence Alignment (ISA) <br> Inferred from Sequence Model (ISM) <br> Inferred from Genomic Context (IGC) <br> Inferred from Reviewed Computational Analysis (RCA)</p>
<p><strong>Author statement evidence codes:</strong> <br> Traceable Author Statement (TAS) <br> Non-traceable Author Statement (NAS)</p>
<p><strong>Curator statement evidence codes:</strong> <br> Inferred by Curator (IC) <br> No biological Data available (ND)</p>
<p><strong>Electronic annotation evidence code:</strong> <br> Inferred from Electronic Annotation (IEA)</p>
</div>
</div>
</section>
<section id="find-genes-based-on-go-info" class="level2">
<h2 class="anchored" data-anchor-id="find-genes-based-on-go-info">Find genes based on GO info</h2>
<p><strong>Find genes matching a regular expression in any associated GO term description:</strong></p>
<div id="cell-80" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>motor_genes_df <span class="op">=</span> gi.get_genes_for_go_regex(<span class="st">'kinesin|dynein'</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>motor_genes_df.head() <span class="op">&lt;&lt;</span> left()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(comment def relationship synonym xref)
geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(comment def relationship synonym xref)</code></pre>
</div>
<div class="cell-output cell-output-display">
<style type="text/css">
#T_41c02 th {
  text-align: left;
}
#T_41c02_row0_col0, #T_41c02_row0_col1, #T_41c02_row0_col2, #T_41c02_row0_col3, #T_41c02_row0_col4, #T_41c02_row1_col0, #T_41c02_row1_col1, #T_41c02_row1_col2, #T_41c02_row1_col3, #T_41c02_row1_col4, #T_41c02_row2_col0, #T_41c02_row2_col1, #T_41c02_row2_col2, #T_41c02_row2_col3, #T_41c02_row2_col4, #T_41c02_row3_col0, #T_41c02_row3_col1, #T_41c02_row3_col2, #T_41c02_row3_col3, #T_41c02_row3_col4, #T_41c02_row4_col0, #T_41c02_row4_col1, #T_41c02_row4_col2, #T_41c02_row4_col3, #T_41c02_row4_col4 {
  text-align: left;
}
</style>

<table id="T_41c02" class="caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_41c02_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">symbol</th>
<th id="T_41c02_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">name</th>
<th id="T_41c02_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">chrom</th>
<th id="T_41c02_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">start</th>
<th id="T_41c02_level0_col4" class="col_heading level0 col4" data-quarto-table-cell-role="th">end</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_41c02_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</td>
<td id="T_41c02_row0_col0" class="data row0 col0">SPAG1</td>
<td id="T_41c02_row0_col1" class="data row0 col1">sperm associated antigen 1</td>
<td id="T_41c02_row0_col2" class="data row0 col2">8</td>
<td id="T_41c02_row0_col3" class="data row0 col3">100158037</td>
<td id="T_41c02_row0_col4" class="data row0 col4">100241903</td>
</tr>
<tr class="even">
<td id="T_41c02_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</td>
<td id="T_41c02_row1_col0" class="data row1 col0">BLOC1S2</td>
<td id="T_41c02_row1_col1" class="data row1 col1">biogenesis of lysosomal organelles complex 1 subunit 2</td>
<td id="T_41c02_row1_col2" class="data row1 col2">10</td>
<td id="T_41c02_row1_col3" class="data row1 col3">100286679</td>
<td id="T_41c02_row1_col4" class="data row1 col4">100273277</td>
</tr>
<tr class="odd">
<td id="T_41c02_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">2</td>
<td id="T_41c02_row2_col0" class="data row2 col0">LZTS2</td>
<td id="T_41c02_row2_col1" class="data row2 col1">leucine zipper tumor suppressor 2</td>
<td id="T_41c02_row2_col2" class="data row2 col2">10</td>
<td id="T_41c02_row2_col3" class="data row2 col3">100996587</td>
<td id="T_41c02_row2_col4" class="data row2 col4">101007832</td>
</tr>
<tr class="even">
<td id="T_41c02_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">3</td>
<td id="T_41c02_row3_col0" class="data row3 col0">DYNC1H1</td>
<td id="T_41c02_row3_col1" class="data row3 col1">dynein cytoplasmic 1 heavy chain 1</td>
<td id="T_41c02_row3_col2" class="data row3 col2">14</td>
<td id="T_41c02_row3_col3" class="data row3 col3">101964572</td>
<td id="T_41c02_row3_col4" class="data row3 col4">102056442</td>
</tr>
<tr class="odd">
<td id="T_41c02_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">4</td>
<td id="T_41c02_row4_col0" class="data row4 col0">KIF1B</td>
<td id="T_41c02_row4_col1" class="data row4 col1">kinesin family member 1B</td>
<td id="T_41c02_row4_col2" class="data row4 col2">1</td>
<td id="T_41c02_row4_col3" class="data row4 col3">10210569</td>
<td id="T_41c02_row4_col4" class="data row4 col4">10381602</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="global-go-enrichment" class="level2">
<h2 class="anchored" data-anchor-id="global-go-enrichment">Global GO enrichment</h2>
<p><strong>Enrichment against all GO terms with all protein coding genes as background:</strong></p>
<div id="cell-83" class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> importlib</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geneinfo</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>importlib.<span class="bu">reload</span>(geneinfo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="54">
<pre><code>&lt;module 'geneinfo' from '/Users/kmt/geneinfo/geneinfo/__init__.py'&gt;</code></pre>
</div>
</div>
<div id="cell-84" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>go_results <span class="op">=</span> gi.go_enrichment(motor_genes_df.symbol)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>go_results.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Could not map gene symbol "KIF28P" to ncbi id</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="55">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">namespace</th>
<th data-quarto-table-cell-role="th">term_id</th>
<th data-quarto-table-cell-role="th">e/p</th>
<th data-quarto-table-cell-role="th">pval_uncorr</th>
<th data-quarto-table-cell-role="th">p_fdr_bh</th>
<th data-quarto-table-cell-role="th">ratio</th>
<th data-quarto-table-cell-role="th">bg_ratio</th>
<th data-quarto-table-cell-role="th">obj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CC</td>
<td>GO:0005874</td>
<td>e</td>
<td>3.174662e-104</td>
<td>5.793757e-101</td>
<td>0.296407</td>
<td>0.014664</td>
<td>&lt;GO:0005874&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>BP</td>
<td>GO:0007018</td>
<td>e</td>
<td>3.469444e-87</td>
<td>4.220231e-83</td>
<td>0.152695</td>
<td>0.002719</td>
<td>&lt;GO:0007018&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>CC</td>
<td>GO:0005871</td>
<td>e</td>
<td>1.323152e-82</td>
<td>1.207376e-79</td>
<td>0.134731</td>
<td>0.002185</td>
<td>&lt;GO:0005871&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>MF</td>
<td>GO:0008017</td>
<td>e</td>
<td>2.849730e-80</td>
<td>1.305176e-76</td>
<td>0.239521</td>
<td>0.012819</td>
<td>&lt;GO:0008017&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>CC</td>
<td>GO:0005813</td>
<td>e</td>
<td>1.960675e-77</td>
<td>1.192744e-74</td>
<td>0.314371</td>
<td>0.030444</td>
<td>&lt;GO:0005813&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>MF</td>
<td>GO:0019894</td>
<td>e</td>
<td>4.661206e-77</td>
<td>1.067416e-73</td>
<td>0.125749</td>
<td>0.002039</td>
<td>&lt;GO:0019894&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>CC</td>
<td>GO:0032391</td>
<td>e</td>
<td>3.269844e-75</td>
<td>1.491866e-72</td>
<td>0.122754</td>
<td>0.001991</td>
<td>&lt;GO:0032391&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>MF</td>
<td>GO:0003777</td>
<td>e</td>
<td>5.259928e-70</td>
<td>8.030157e-67</td>
<td>0.125749</td>
<td>0.002331</td>
<td>&lt;GO:0003777&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>MF</td>
<td>GO:0045505</td>
<td>e</td>
<td>7.657838e-68</td>
<td>8.768225e-65</td>
<td>0.110778</td>
<td>0.001797</td>
<td>&lt;GO:0045505&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>CC</td>
<td>GO:0016235</td>
<td>e</td>
<td>3.632820e-64</td>
<td>1.325979e-61</td>
<td>0.104790</td>
<td>0.001699</td>
<td>&lt;GO:0016235&gt;</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><strong>Show GO DAG for enrichment results:</strong></p>
<div id="cell-86" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_enrichment_results(go_results.obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="56">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-47-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Same but only for top 10 GO terms:</strong></p>
<div id="cell-88" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_enrichment_results(go_results.obj[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="57">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-48-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Descriptions of the same 10 GO terms:</strong></p>
<div id="cell-90" class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>gi.go_info(go_results.term_id[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005874:</span></strong> <strong>microtubule</strong><br>
Any of the long, generally straight, hollow tubes of internal diameter 12-15 nm and external diameter 24 nm found in a wide variety of eukaryotic cells; each consists (usually) of 13 protofilaments of polymeric tubulin, staggered in such a manner that the tubulin monomers are arranged in a helical pattern on the microtubular surface, and with the alpha/beta axes of the tubulin subunits parallel to the long axis of the tubule; exist in equilibrium with pool of tubulin monomers and can be rapidly assembled or disassembled in response to physiological stimuli; concerned with force generation, e.g.&nbsp;in the spindle.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0007018:</span></strong> <strong>microtubule-based movement</strong><br>
A microtubule-based process that results in the movement of organelles, other microtubules, or other cellular components. Examples include motor-driven movement along microtubules and movement driven by polymerization or depolymerization of microtubules.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005871:</span></strong> <strong>kinesin complex</strong><br>
Any complex that includes a dimer of molecules from the kinesin superfamily, a group of related proteins that contain an extended region of predicted alpha-helical coiled coil in the main chain that likely produces dimerization. The native complexes of several kinesin family members have also been shown to contain additional peptides, often designated light chains as all of the noncatalytic subunits that are currently known are smaller than the chain that contains the motor unit. Kinesin complexes generally possess a force-generating enzymatic activity, or motor, which converts the free energy of the gamma phosphate bond of ATP into mechanical work.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0008017:</span></strong> <strong>microtubule binding</strong><br>
Binding to a microtubule, a filament composed of tubulin monomers.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0005813:</span></strong> <strong>centrosome</strong><br>
A structure comprised of a core structure (in most organisms, a pair of centrioles) and peripheral material from which a microtubule-based structure, such as a spindle apparatus, is organized. Centrosomes occur close to the nucleus during interphase in many eukaryotic cells, though in animal cells it changes continually during the cell-division cycle.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0019894:</span></strong> <strong>kinesin binding</strong><br>
Interacting selectively and non-covalently and stoichiometrically with kinesin, a member of a superfamily of microtubule-based motor proteins that perform force-generating tasks such as organelle transport and chromosome segregation.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0032391:</span></strong> <strong>photoreceptor connecting cilium</strong><br>
The portion of the photoreceptor cell cilium linking the photoreceptor inner and outer segments. Itâ€™s considered to be equivalent to the ciliary transition zone.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0003777:</span></strong> <strong>microtubule motor activity</strong><br>
A motor activity that generates movement along a microtubule, driven by ATP hydrolysis.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0045505:</span></strong> <strong>dynein intermediate chain binding</strong><br>
Binding to an intermediate chain of the dynein complex.</p>
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p><strong><span style="color:gray;">GO:0016235:</span></strong> <strong>aggresome</strong><br>
An inclusion body formed by dynein-dependent retrograde transport of an aggregated protein on microtubules.</p>
<hr>
</div>
</div>
<p><strong>Show the GO DAG for a set of GO terms and their parent terms:</strong></p>
<div id="cell-92" class="cell" data-execution_count="59">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_terms(go_results.term_id[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="59">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-50-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Specifying a custom set of background genes:</strong></p>
<p>The query genes should be a subset of the background genes:</p>
<div id="cell-94" class="cell" data-execution_count="60">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>all_protein_coding <span class="op">=</span> <span class="bu">set</span>(pd.read_csv(<span class="st">'geneinfo_cache/9606_protein_genes.txt'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>).Symbol)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co"># only the first 50 genes:</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>subset_background <span class="op">=</span> <span class="bu">list</span>(all_protein_coding)[:<span class="dv">50</span>] <span class="op">+</span> motor_genes_df.symbol.tolist()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-95" class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>go_results <span class="op">=</span> gi.go_enrichment(motor_genes_df.symbol, background_genes<span class="op">=</span>subset_background)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>go_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Could not map gene symbol "KIF28P" to ncbi id
Could not map gene symbol "KIF28P" to ncbi id</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="61">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">namespace</th>
<th data-quarto-table-cell-role="th">term_id</th>
<th data-quarto-table-cell-role="th">e/p</th>
<th data-quarto-table-cell-role="th">pval_uncorr</th>
<th data-quarto-table-cell-role="th">p_fdr_bh</th>
<th data-quarto-table-cell-role="th">ratio</th>
<th data-quarto-table-cell-role="th">bg_ratio</th>
<th data-quarto-table-cell-role="th">obj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CC</td>
<td>GO:0005813</td>
<td>e</td>
<td>8.279747e-07</td>
<td>0.000378</td>
<td>0.314371</td>
<td>0.276042</td>
<td>&lt;GO:0005813&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CC</td>
<td>GO:0005874</td>
<td>e</td>
<td>2.975553e-06</td>
<td>0.000678</td>
<td>0.296407</td>
<td>0.260417</td>
<td>&lt;GO:0005874&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>CC</td>
<td>GO:0005737</td>
<td>e</td>
<td>2.878076e-05</td>
<td>0.004375</td>
<td>0.538922</td>
<td>0.497396</td>
<td>&lt;GO:0005737&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>MF</td>
<td>GO:0000981</td>
<td>p</td>
<td>9.358545e-05</td>
<td>0.018951</td>
<td>0.011976</td>
<td>0.028646</td>
<td>&lt;GO:0000981&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>MF</td>
<td>GO:0008017</td>
<td>e</td>
<td>6.502644e-05</td>
<td>0.018951</td>
<td>0.239521</td>
<td>0.210938</td>
<td>&lt;GO:0008017&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>CC</td>
<td>GO:0005929</td>
<td>e</td>
<td>3.614207e-04</td>
<td>0.032962</td>
<td>0.164671</td>
<td>0.143229</td>
<td>&lt;GO:0005929&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>CC</td>
<td>GO:0036064</td>
<td>e</td>
<td>3.598807e-04</td>
<td>0.032962</td>
<td>0.167665</td>
<td>0.145833</td>
<td>&lt;GO:0036064&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>CC</td>
<td>GO:0005930</td>
<td>e</td>
<td>6.152330e-04</td>
<td>0.046758</td>
<td>0.155689</td>
<td>0.135417</td>
<td>&lt;GO:0005930&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>CC</td>
<td>GO:0005829</td>
<td>e</td>
<td>7.499627e-04</td>
<td>0.048855</td>
<td>0.494012</td>
<td>0.460938</td>
<td>&lt;GO:0005829&gt;</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><strong>Use only the X chromosome as background:</strong></p>
<div id="cell-97" class="cell" data-execution_count="62">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>go_results <span class="op">=</span> gi.go_enrichment(motor_genes_df.symbol, background_chrom<span class="op">=</span><span class="st">'X'</span>)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>go_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Could not map gene symbol "KIF28P" to ncbi id</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="62">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">namespace</th>
<th data-quarto-table-cell-role="th">term_id</th>
<th data-quarto-table-cell-role="th">e/p</th>
<th data-quarto-table-cell-role="th">pval_uncorr</th>
<th data-quarto-table-cell-role="th">p_fdr_bh</th>
<th data-quarto-table-cell-role="th">ratio</th>
<th data-quarto-table-cell-role="th">bg_ratio</th>
<th data-quarto-table-cell-role="th">obj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>MF</td>
<td>GO:0045505</td>
<td>e</td>
<td>0.000041</td>
<td>0.010017</td>
<td>0.333333</td>
<td>0.002345</td>
<td>&lt;GO:0045505&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>MF</td>
<td>GO:0070840</td>
<td>e</td>
<td>0.000041</td>
<td>0.010017</td>
<td>0.333333</td>
<td>0.002345</td>
<td>&lt;GO:0070840&gt;</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>MF</td>
<td>GO:0008017</td>
<td>e</td>
<td>0.000031</td>
<td>0.010017</td>
<td>0.500000</td>
<td>0.012896</td>
<td>&lt;GO:0008017&gt;</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>CC</td>
<td>GO:0016235</td>
<td>e</td>
<td>0.000041</td>
<td>0.020475</td>
<td>0.333333</td>
<td>0.002345</td>
<td>&lt;GO:0016235&gt;</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="enrichment-among-a-subset-of-go-terms" class="level2">
<h2 class="anchored" data-anchor-id="enrichment-among-a-subset-of-go-terms">Enrichment among a subset of GO terms</h2>
<p>Extract terms matching regular expression:</p>
<div id="cell-99" class="cell" data-execution_count="63">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">#target_terms = gi.get_terms_for_go_regex('brain')</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co">#target_terms = gi.get_terms_for_go_regex('microtubul')</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co"># target_terms = gi.get_terms_for_go_regex('spindle')</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>target_terms <span class="op">=</span> gi.get_terms_for_go_regex(<span class="vs">r'neuron\S* migration'</span>)</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_terms(target_terms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(comment def relationship synonym xref)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="63">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-54-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-100" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>go_results <span class="op">=</span> gi.go_enrichment(motor_genes_df.symbol, terms<span class="op">=</span>target_terms, list_study_genes<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>go_results.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(def relationship)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Could not map gene symbol "KIF28P" to ncbi id</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="64">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">namespace</th>
<th data-quarto-table-cell-role="th">term_id</th>
<th data-quarto-table-cell-role="th">e/p</th>
<th data-quarto-table-cell-role="th">pval_uncorr</th>
<th data-quarto-table-cell-role="th">p_fdr_bh</th>
<th data-quarto-table-cell-role="th">ratio</th>
<th data-quarto-table-cell-role="th">bg_ratio</th>
<th data-quarto-table-cell-role="th">obj</th>
<th data-quarto-table-cell-role="th">study_genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CC</td>
<td>GO:0005615</td>
<td>p</td>
<td>0.000016</td>
<td>0.000063</td>
<td>0.020958</td>
<td>0.076426</td>
<td>&lt;GO:0005615&gt;</td>
<td>[ACTB, BLOC1S1, HSPA1A, SNCA, LRP8, DYNC2I1, T...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CC</td>
<td>GO:0032991</td>
<td>e</td>
<td>0.000155</td>
<td>0.000310</td>
<td>0.071856</td>
<td>0.031027</td>
<td>&lt;GO:0032991&gt;</td>
<td>[ACTB, HTT, HNRNPU, HSPA1A, HSPA1B, HSP90AB1, ...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>BP</td>
<td>GO:0140650</td>
<td>e</td>
<td>0.000778</td>
<td>0.031519</td>
<td>0.005988</td>
<td>0.000146</td>
<td>&lt;GO:0140650&gt;</td>
<td>[PAFAH1B1, NDEL1]</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>BP</td>
<td>GO:0001764</td>
<td>e</td>
<td>0.000515</td>
<td>0.031519</td>
<td>0.023952</td>
<td>0.005487</td>
<td>&lt;GO:0001764&gt;</td>
<td>[BBS4, PAFAH1B1, PCM1, PSEN1, DISC1, DCDC2, ND...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-101" class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_enrichment_results(go_results.obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="65">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-56-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Extract terms matching two regular expressions in any order:</p>
<div id="cell-103" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>pattern_one <span class="op">=</span> <span class="vs">r'microtub\S*'</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>pattern_two <span class="op">=</span> <span class="vs">r'spindle'</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>regex <span class="op">=</span> <span class="vs">fr"(?=.*(</span><span class="sc">{</span>pattern_one<span class="sc">}</span><span class="vs">))(?=.*(</span><span class="sc">{</span>pattern_two<span class="sc">}</span><span class="vs">))(?=.*)"</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>target_terms <span class="op">=</span> gi.get_terms_for_go_regex(regex)</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_for_terms(target_terms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(comment def relationship synonym xref)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="66">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-57-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-104" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>go_results <span class="op">=</span> gi.go_enrichment(motor_genes_df.symbol, terms<span class="op">=</span>target_terms, list_study_genes<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>go_results.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>geneinfo_cache/go-basic.obo: fmt(1.2) rel(2024-09-08) 44,296 Terms; optional_attrs(def relationship)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Could not map gene symbol "KIF28P" to ncbi id</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="67">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">namespace</th>
<th data-quarto-table-cell-role="th">term_id</th>
<th data-quarto-table-cell-role="th">e/p</th>
<th data-quarto-table-cell-role="th">pval_uncorr</th>
<th data-quarto-table-cell-role="th">p_fdr_bh</th>
<th data-quarto-table-cell-role="th">ratio</th>
<th data-quarto-table-cell-role="th">bg_ratio</th>
<th data-quarto-table-cell-role="th">obj</th>
<th data-quarto-table-cell-role="th">study_genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CC</td>
<td>GO:0005874</td>
<td>e</td>
<td>3.174662e-104</td>
<td>1.428598e-102</td>
<td>0.296407</td>
<td>0.014664</td>
<td>&lt;GO:0005874&gt;</td>
<td>[APC, KIF1A, CENPE, DCTN1, DNAH5, DNAH6, DNAH8...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>BP</td>
<td>GO:0007018</td>
<td>e</td>
<td>3.469444e-87</td>
<td>3.816388e-85</td>
<td>0.152695</td>
<td>0.002719</td>
<td>&lt;GO:0007018&gt;</td>
<td>[CENPE, DYNC1I2, DYNC1LI2, KIF2A, KIF3C, KIF5A...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>CC</td>
<td>GO:0005813</td>
<td>e</td>
<td>1.960675e-77</td>
<td>4.411518e-76</td>
<td>0.314371</td>
<td>0.030444</td>
<td>&lt;GO:0005813&gt;</td>
<td>[APC, ARL3, BBS4, BICD1, CENPF, CETN1, CETN2, ...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>CC</td>
<td>GO:0005868</td>
<td>e</td>
<td>3.150407e-42</td>
<td>4.725610e-41</td>
<td>0.068862</td>
<td>0.001117</td>
<td>&lt;GO:0005868&gt;</td>
<td>[DYNC1H1, DYNC1I1, DYNC1I2, DYNC1LI2, DYNLT3, ...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>CC</td>
<td>GO:0030286</td>
<td>e</td>
<td>5.844463e-35</td>
<td>6.575021e-34</td>
<td>0.056886</td>
<td>0.000923</td>
<td>&lt;GO:0030286&gt;</td>
<td>[DCTN1, DNAH6, DNAH9, DYNC1H1, DYNC1I2, DYNC1L...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-105" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>gi.show_go_dag_enrichment_results(go_results.obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="68">
<div>
<figure class="figure">
<p><img src="examples_files/figure-html/cell-59-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/munch-group\.github\.io\/geneinfo\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../pages/getting_started.html" class="pagination-link" aria-label="Overview">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Overview</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../api/" class="pagination-link" aria-label="API reference">
        <span class="nav-page-text">API reference</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>